{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "#%%appyter init\n",
    "from appyter import magic\n",
    "magic.init(lambda _=globals: _())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter hide_code\n",
    "\n",
    "{% do SectionField(\n",
    "    name='primary',\n",
    "    title='Upload Tumor Expression',\n",
    "    img='upload.png'\n",
    ") %}\n",
    "\n",
    "{% set data_type = TabField(\n",
    "    name='data_type',\n",
    "    label='Data Type',\n",
    "    default='scRNA-seq',\n",
    "    description='Start with either scRNA-seq or bulk RNA-seq data. If bulk RNA-seq data is selected we will utilize pre-curated reference matrices to identify cell type-specific expression vectors.',\n",
    "    required=True,\n",
    "    section='primary',\n",
    "    choices={\n",
    "        'scRNA-seq': [\n",
    "            TabField(\n",
    "            name='sc_data_type',\n",
    "            label='Data Format (Control Profile)',\n",
    "            default='Plain Text',\n",
    "            description='Choose a format for scRNA-seq data',\n",
    "            required=True,\n",
    "            choices= {\n",
    "                'Plain Text': [\n",
    "                        CustomFileField(\n",
    "                        name='sc_rna_file_ctrl',\n",
    "                        label='control scRNA-seq',\n",
    "                        description='''\n",
    "    File should be a tsv/csv of the form:\n",
    "\n",
    "                                        cell 1   cell 2    ...\n",
    "                    ------------------------------\n",
    "            Gene/Protein 1    0          2       ...\n",
    "                    ------------------------------\n",
    "            Gene/Protein 2    1         1        ...\n",
    "                    ------------------------------\n",
    "                        ...                       ... \n",
    "                        ''',\n",
    "                        default='data/sarcoma-v2-proteome-SpectrumMill-ratio-QCfilter-NArm.tsv',\n",
    "                        required=False,\n",
    "                        examples={\n",
    "                            'BR_proteomics_tumor.tsv': 'https://s3.amazonaws.com/multiomics2paper/public/BR_proteomics_tumor.tsv',\n",
    "                        },\n",
    "                    ),\n",
    "                    CustomFileField(\n",
    "                        name='sc_meta_file',\n",
    "                        label='metadata',\n",
    "                        description='''\n",
    "                        File should be a tsv/csv of the form:\n",
    "\n",
    "                            Patient 1 Tumor    Patient 2 Tumor  ...\n",
    "                        ---------------------------------------\n",
    "                        Gene/Protein 1    0                   200       ...\n",
    "                        ---------------------------------------\n",
    "                        Gene/Protein 2    5                   180       ...\n",
    "                        ---------------------------------------\n",
    "                        ...                       ...                    ...        ...\n",
    "                        ''',\n",
    "                        default='data/sarcoma-v2-proteome-SpectrumMill-ratio-QCfilter-NArm.tsv',\n",
    "                        required=False,\n",
    "                        examples={\n",
    "                            'BR_proteomics_tumor.tsv': 'https://s3.amazonaws.com/multiomics2paper/public/BR_proteomics_tumor.tsv',\n",
    "                        },\n",
    "                    ),\n",
    "                    TextField(\n",
    "                        name='cell_type_col',\n",
    "                        label='Cell Type Column (Optional)',\n",
    "                        description='Name of cell type column in uploaded metadata (if not selected, cell type identification will be performed)',\n",
    "                        default='cell_type',\n",
    "                        required=False\n",
    "                    )\n",
    "                    ],\n",
    "                '.mtx': [\n",
    "                        CustomFileField(\n",
    "                        name='sc_rna_file',\n",
    "                        label='scRNA-seq',\n",
    "                        description='''\n",
    "    File should be a tsv/csv of the form:\n",
    "\n",
    "                                        cell 1   cell 2    ...\n",
    "                    ------------------------------\n",
    "            Gene/Protein 1    0          2       ...\n",
    "                    ------------------------------\n",
    "            Gene/Protein 2    1         1        ...\n",
    "                    ------------------------------\n",
    "                        ...                       ... \n",
    "                        ''',\n",
    "                        default='data/sarcoma-v2-proteome-SpectrumMill-ratio-QCfilter-NArm.tsv',\n",
    "                        required=False,\n",
    "                        examples={\n",
    "                            'BR_proteomics_tumor.tsv': 'https://s3.amazonaws.com/multiomics2paper/public/BR_proteomics_tumor.tsv',\n",
    "                        },\n",
    "                    ),\n",
    "                    CustomFileField(\n",
    "                        name='sc_meta_file',\n",
    "                        label='metadata',\n",
    "                        description='''\n",
    "                        File should be a tsv/csv of the form:\n",
    "\n",
    "                            Patient 1 Tumor    Patient 2 Tumor  ...\n",
    "                        ---------------------------------------\n",
    "                        Gene/Protein 1    0                   200       ...\n",
    "                        ---------------------------------------\n",
    "                        Gene/Protein 2    5                   180       ...\n",
    "                        ---------------------------------------\n",
    "                        ...                       ...                    ...        ...\n",
    "                        ''',\n",
    "                        default='data/sarcoma-v2-proteome-SpectrumMill-ratio-QCfilter-NArm.tsv',\n",
    "                        required=False,\n",
    "                        examples={\n",
    "                            'BR_proteomics_tumor.tsv': 'https://s3.amazonaws.com/multiomics2paper/public/BR_proteomics_tumor.tsv',\n",
    "                        },\n",
    "                    )]\n",
    "            }\n",
    "        ),\n",
    "        TabField(\n",
    "            name='sc_data_type',\n",
    "            label='Data format',\n",
    "            default='Plain Text',\n",
    "            description='Choose a format for scRNA-seq data',\n",
    "            required=True,\n",
    "            choices= {\n",
    "                'Plain Text': [\n",
    "                        CustomFileField(\n",
    "                        name='sc_rna_file',\n",
    "                        label='scRNA-seq',\n",
    "                        description='''\n",
    "    File should be a tsv/csv of the form:\n",
    "\n",
    "                                        cell 1   cell 2    ...\n",
    "                    ------------------------------\n",
    "            Gene/Protein 1    0          2       ...\n",
    "                    ------------------------------\n",
    "            Gene/Protein 2    1         1        ...\n",
    "                    ------------------------------\n",
    "                        ...                       ... \n",
    "                        ''',\n",
    "                        default='data/sarcoma-v2-proteome-SpectrumMill-ratio-QCfilter-NArm.tsv',\n",
    "                        required=False,\n",
    "                        examples={\n",
    "                            'BR_proteomics_tumor.tsv': 'https://s3.amazonaws.com/multiomics2paper/public/BR_proteomics_tumor.tsv',\n",
    "                        },\n",
    "                    ),\n",
    "                    CustomFileField(\n",
    "                        name='sc_meta_file',\n",
    "                        label='metadata',\n",
    "                        description='''\n",
    "                        File should be a tsv/csv of the form:\n",
    "\n",
    "                            Patient 1 Tumor    Patient 2 Tumor  ...\n",
    "                        ---------------------------------------\n",
    "                        Gene/Protein 1    0                   200       ...\n",
    "                        ---------------------------------------\n",
    "                        Gene/Protein 2    5                   180       ...\n",
    "                        ---------------------------------------\n",
    "                        ...                       ...                    ...        ...\n",
    "                        ''',\n",
    "                        default='data/sarcoma-v2-proteome-SpectrumMill-ratio-QCfilter-NArm.tsv',\n",
    "                        required=False,\n",
    "                        examples={\n",
    "                            'BR_proteomics_tumor.tsv': 'https://s3.amazonaws.com/multiomics2paper/public/BR_proteomics_tumor.tsv',\n",
    "                        },\n",
    "                    ),\n",
    "                    TextField(\n",
    "                        name='condition name',\n",
    "                        label='condition name',\n",
    "                        description='Name of the condition',\n",
    "                        default='Tumor',\n",
    "                        required=True\n",
    "                    )\n",
    "                    ],\n",
    "                '.mtx': [\n",
    "                        CustomFileField(\n",
    "                        name='sc_rna_file',\n",
    "                        label='scRNA-seq',\n",
    "                        description='''\n",
    "    File should be a tsv/csv of the form:\n",
    "\n",
    "                                        cell 1   cell 2    ...\n",
    "                    ------------------------------\n",
    "            Gene/Protein 1    0          2       ...\n",
    "                    ------------------------------\n",
    "            Gene/Protein 2    1         1        ...\n",
    "                    ------------------------------\n",
    "                        ...                       ... \n",
    "                        ''',\n",
    "                        default='data/sarcoma-v2-proteome-SpectrumMill-ratio-QCfilter-NArm.tsv',\n",
    "                        required=False,\n",
    "                        examples={\n",
    "                            'BR_proteomics_tumor.tsv': 'https://s3.amazonaws.com/multiomics2paper/public/BR_proteomics_tumor.tsv',\n",
    "                        },\n",
    "                    ),\n",
    "                    CustomFileField(\n",
    "                        name='sc_meta_file',\n",
    "                        label='metadata',\n",
    "                        description='''\n",
    "                        File should be a tsv/csv of the form:\n",
    "\n",
    "                            Patient 1 Tumor    Patient 2 Tumor  ...\n",
    "                        ---------------------------------------\n",
    "                        Gene/Protein 1    0                   200       ...\n",
    "                        ---------------------------------------\n",
    "                        Gene/Protein 2    5                   180       ...\n",
    "                        ---------------------------------------\n",
    "                        ...                       ...                    ...        ...\n",
    "                        ''',\n",
    "                        default='data/sarcoma-v2-proteome-SpectrumMill-ratio-QCfilter-NArm.tsv',\n",
    "                        required=False,\n",
    "                        examples={\n",
    "                            'BR_proteomics_tumor.tsv': 'https://s3.amazonaws.com/multiomics2paper/public/BR_proteomics_tumor.tsv',\n",
    "                        },\n",
    "                    )]\n",
    "            }\n",
    "        )],\n",
    "        'bulk RNA-seq':[\n",
    "            CustomFileField(\n",
    "                name=' bulk_expr',\n",
    "                label='RNA-seq expression',\n",
    "                description='''\n",
    "                File should be a tsv/csv of the form:\n",
    "\n",
    "                    Patient 1 Tumor    Patient 2 Tumor  ...\n",
    "                ---------------------------------------\n",
    "                Gene/Protein 1    0                   200       ...\n",
    "                ---------------------------------------\n",
    "                Gene/Protein 2    5                   180       ...\n",
    "                ---------------------------------------\n",
    "                ...                       ...                    ...        ...\n",
    "                ''',\n",
    "                default='data/sarcoma-v2-proteome-SpectrumMill-ratio-QCfilter-NArm.tsv',\n",
    "                required=False,\n",
    "                examples={\n",
    "                    'BR_proteomics_tumor.tsv': 'https://s3.amazonaws.com/multiomics2paper/public/BR_proteomics_tumor.tsv',\n",
    "                },\n",
    "                section='primary',\n",
    "            ),\n",
    "            CustomFileField(\n",
    "                        name='sc_meta_file',\n",
    "                        label='metadata',\n",
    "                        description='''\n",
    "                        File should be a tsv/csv of the form:\n",
    "\n",
    "                            Patient 1 Tumor    Patient 2 Tumor  ...\n",
    "                        ---------------------------------------\n",
    "                        Gene/Protein 1    0                   200       ...\n",
    "                        ---------------------------------------\n",
    "                        Gene/Protein 2    5                   180       ...\n",
    "                        ---------------------------------------\n",
    "                        ...                       ...                    ...        ...\n",
    "                        ''',\n",
    "                        default='data/sarcoma-v2-proteome-SpectrumMill-ratio-QCfilter-NArm.tsv',\n",
    "                        required=False,\n",
    "                        examples={\n",
    "                            'BR_proteomics_tumor.tsv': 'https://s3.amazonaws.com/multiomics2paper/public/BR_proteomics_tumor.tsv',\n",
    "                        },\n",
    "                    ),\n",
    "                    TextField(\n",
    "                        name='condition_col',\n",
    "                        label='Condition Column',\n",
    "                        description='Name of the condition column in uploaded metadata',\n",
    "                        default='Condition',\n",
    "                        required=True\n",
    "                    ),\n",
    "                    TextField(\n",
    "                        name='ctrl_condition',\n",
    "                        label='Control Condition Name',\n",
    "                        description='Name of the control condition in the condition column in uploaded metadata',\n",
    "                        default='healthy',\n",
    "                        required=True\n",
    "                    ),\n",
    "                    ChoiceField(\n",
    "                        name='reference',\n",
    "                        label='Single Cell Reference',\n",
    "                        description='Choose a single cell reference to use for cell type deconvolution',\n",
    "                        choices={\n",
    "                            \"Pancreas - Tabula Sapiens (Human)\": \"https://cellxgene.cziscience.com/collections/e5f58829-1a66-40b5-a624-9046778e74f5\",\n",
    "                            \"Pancreas - Tritschler et al. (Mouse)\": \"https://cellxgene.cziscience.com/collections/0a77d4c0-d5d0-40f0-aa1a-5e1429bcbd7e\",\n",
    "                            \"Adipose Tissue - Tabula Sapiens (Human)\": \"https://cellxgene.cziscience.com/collections/e5f58829-1a66-40b5-a624-9046778e74f5\",\n",
    "                            \"Adipose Tissue - Emont et al. (Mouse)\": \"https://cellxgene.cziscience.com/collections/fe0e718d-2ee9-42cc-894b-0b490f437dfd\",\n",
    "                            \"Adipose Tissue - Tabula Muris (Mouse)\": \"https://cellxgene.cziscience.com/collections/0b9d8a04-bb9d-44da-aa27-705bb65b54eb\",\n",
    "                            \"Liver - MacParland et al. 2018 (Human)\": \"https://cellxgene.cziscience.com/collections/bd5230f4-cd76-4d35-9ee5-89b3e7475659\",\n",
    "                            \"Blood - Tabula Sapiens (Human)\": \"https://cellxgene.cziscience.com/collections/e5f58829-1a66-40b5-a624-9046778e74f5\",\n",
    "                            \"Kidney - Tabula Sapiens(Human)\": \"https://cellxgene.cziscience.com/collections/e5f58829-1a66-40b5-a624-9046778e74f5\",\n",
    "                            \"Brain - Tabula Muris (Mouse)\": \"https://cellxgene.cziscience.com/collections/0b9d8a04-bb9d-44da-aa27-705bb65b54eb\",\n",
    "                            \"Skeletal Muscle - Domínguez Conde et al. 2022 (Human)\": \"https://cellxgene.cziscience.com/collections/62ef75e4-cbea-454e-a0ce-998ec40223d3\",\n",
    "                            \"Heart - Tabula Sapiens (Human)\": \"https://cellxgene.cziscience.com/collections/e5f58829-1a66-40b5-a624-9046778e74f5\",\n",
    "                            \"Heart - Litviňuková et al. 2020 (Mouse)\": \"https://cellxgene.cziscience.com/collections/b52eb423-5d0d-4645-b217-e1c6d38b2e72\",\n",
    "                            \"Heart - Tabula Muris (Mouse)\": \"https://cellxgene.cziscience.com/collections/0b9d8a04-bb9d-44da-aa27-705bb65b54eb\",\n",
    "                            \"Muscle - Tabula Sapiens (Human)\": \"https://cellxgene.cziscience.com/collections/e5f58829-1a66-40b5-a624-9046778e74f5\",\n",
    "                            \"Muscle - Tabula Muris (Mouse)\": \"https://cellxgene.cziscience.com/collections/0b9d8a04-bb9d-44da-aa27-705bb65b54eb\",\n",
    "                            \"Skin - Tabula Sapiens (Human)\": \"https://cellxgene.cziscience.com/collections/e5f58829-1a66-40b5-a624-9046778e74f5\",\n",
    "                            \"Retina - Li et al. 2023 (Human)\": \"https://cellxgene.cziscience.com/collections/4c6eaf5c-6d57-4c76-b1e9-60df8c655f1e\",\n",
    "                            \"Retina - Cowan et al. 2020 (Mouse)\": \"https://cellxgene.cziscience.com/collections/2f4c738f-e2f3-4553-9db2-0582a38ea4dc\",\n",
    "                            \"Breast - Reed et al. 2024 (Human)\": \"https://cellxgene.cziscience.com/collections/48259aa8-f168-4bf5-b797-af8e88da6637\",\n",
    "                            \"Lung - Tabula Sapiens (Human)\": \"https://cellxgene.cziscience.com/collections/e5f58829-1a66-40b5-a624-9046778e74f5\",\n",
    "                            \"Lung - Tabula Muris (Mouse)\": \"https://cellxgene.cziscience.com/collections/0b9d8a04-bb9d-44da-aa27-705bb65b54eb\",\n",
    "                            \"Uterus - Tabula Muris (Mouse)\": \"https://cellxgene.cziscience.com/collections/0b9d8a04-bb9d-44da-aa27-705bb65b54eb\",\n",
    "                            \"Bladder - Tabula Sapiens (Human)\": \"https://cellxgene.cziscience.com/collections/0b9d8a04-bb9d-44da-aa27-705bb65b54eb\",\n",
    "                            \"Bladder - Tabula Muris (Mouse)\": \"https://cellxgene.cziscience.com/collections/0b9d8a04-bb9d-44da-aa27-705bb65b54eb\",\n",
    "                            \"Gingiva - Easter et al. 2024 (Human)\": \"https://cellxgene.cziscience.com/collections/71f4bccf-53d4-4c12-9e80-e73bfb89e398\",\n",
    "                            \"Gingiva - Easter et al. 2024 (Mouse)\": \"https://cellxgene.cziscience.com/collections/67ba665e-0611-4b53-a522-40c2e0dc6df7\",\n",
    "                            \"Spleen - Tabula Muris (Mouse)\": \"https://cellxgene.cziscience.com/collections/0b9d8a04-bb9d-44da-aa27-705bb65b54eb\"\n",
    "                            }\n",
    "                    )\n",
    "         ]\n",
    "        }\n",
    ") %}\n",
    "\n",
    "\n",
    "{% set n_neighbors = IntField(\n",
    "    name='n_neighbors',\n",
    "    label='N neighbors', \n",
    "    default=15, \n",
    "    min=2,\n",
    "    max=100,\n",
    "    step=1,\n",
    "    description='The size of local neighborhood (in terms of number of neighboring data points) used for manifold approximation.', \n",
    "    section='primary'\n",
    ")\n",
    "%}\n",
    "\n",
    "{% set min_dist = FloatField(\n",
    "    name='min_dist',\n",
    "    label='Min Distance', \n",
    "    default=0.01, \n",
    "    min=0.00001,\n",
    "    max=0.9,\n",
    "    step=0.00001,\n",
    "    description='The effective minimum distance between embedded points.', \n",
    "    section='primary'\n",
    ")\n",
    "%}\n",
    "\n",
    "{% set resolution = FloatField(\n",
    "    name='resolution',\n",
    "    label='Resolution', \n",
    "    default=1, \n",
    "    min=0.1,\n",
    "    max=2,\n",
    "    step=0.1,\n",
    "    description='A parameter value controlling the coarseness of the clustering. Higher values lead to more clusters.', \n",
    "    section='primary'\n",
    ")\n",
    "%}\n",
    "\n",
    "\n",
    "{% set membrane_screener_list = 'https://appyters.maayanlab.cloud/storage/Tumor_Gene_Target_Screener/surfaceome.csv' %}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "from helpers import *\n",
    "import os\n",
    "import re\n",
    "import random\n",
    "import qnorm\n",
    "from string import ascii_uppercase\n",
    "from tqdm import tqdm\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "\n",
    "from matplotlib import cm\n",
    "from matplotlib.patches import Rectangle\n",
    "from matplotlib.patches import Patch\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from matplotlib.colors import LinearSegmentedColormap\n",
    "%matplotlib inline\n",
    "\n",
    "import networkx as nx\n",
    "\n",
    "from IPython.display import HTML, display, Markdown, FileLink\n",
    "from matplotlib.gridspec import GridSpec\n",
    "from matplotlib_venn import venn2\n",
    "from maayanlab_bioinformatics.normalization import zscore_normalize, log2_normalize\n",
    "from scipy.stats import zscore\n",
    "from scipy.stats import ttest_ind, ttest_1samp, ttest_ind_from_stats\n",
    "from maayanlab_bioinformatics.dge import limma_voom_differential_expression\n",
    "from maayanlab_bioinformatics.harmonization.ncbi_genes import ncbi_genes_lookup\n",
    "from maayanlab_bioinformatics.api import enrichr_link_from_genes\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "\n",
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "\n",
    "os.makedirs('results', exist_ok=True)\n",
    "os.makedirs('figures', exist_ok=True)\n",
    "try:\n",
    "    lookup = ncbi_genes_lookup(organism='Mammalia/Homo_sapiens')\n",
    "except:\n",
    "    import urllib.request, json \n",
    "    with urllib.request.urlopen(\"https://s3.amazonaws.com/multiomics2paper/public/ncbi_genes_disambiguated.json\") as url:\n",
    "        ncbi_genes_disambiguated = json.load(url)\n",
    "        lookup = ncbi_genes_disambiguated.get\n",
    "    \n",
    "fig_counter = 2\n",
    "table_counter = 1\n",
    "letter_counter = 0\n",
    "discussion_results = {}\n",
    "\n",
    "import sys\n",
    "import contextlib\n",
    "@contextlib.contextmanager\n",
    "def suppress_output(stdout=True, stderr=True, dest=os.devnull):\n",
    "    ''' Usage:\n",
    "    with suppress_output():\n",
    "        print('hi')\n",
    "    '''\n",
    "    dev_null = open(dest, 'a')\n",
    "    if stdout:\n",
    "        _stdout = sys.stdout\n",
    "        sys.stdout = dev_null\n",
    "    if stderr:\n",
    "        _stderr = sys.stderr\n",
    "        sys.stderr = dev_null\n",
    "    try:\n",
    "        yield\n",
    "    finally:\n",
    "        if stdout:\n",
    "            sys.stdout = _stdout\n",
    "        if stderr:\n",
    "            sys.stderr = _stderr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if has_rna_file:\n",
    "    cluster_dges = {}\n",
    "    dge_table = {}\n",
    "    for c in sorted(clusters):\n",
    "        cluster_dges[c] = {}\n",
    "        cluster_samples = list(cluster_df[cluster_df['cluster'] == c].index.values)\n",
    "        other_samples = [s for s in rna_df.columns if s not in cluster_samples]\n",
    "        if rna_df.min().min() < 0:\n",
    "            rna_df_dge = rna_df + abs(rna_df.min().min())\n",
    "        else:\n",
    "            rna_df_dge = rna_df\n",
    "        with suppress_output():\n",
    "            dge = limma_voom_differential_expression(\n",
    "                rna_df_dge[other_samples], rna_df_dge[cluster_samples],\n",
    "                voom_design=True,\n",
    "            )\n",
    "        dge_filename = f'results/cluster_{c}_vs_rest_dge_limma.tsv'\n",
    "        dge.sort_values(by=\"adj.P.Val\", ascending=True).to_csv(dge_filename, sep='\\t')\n",
    "        sig_dge = dge[dge['adj.P.Val'] < 0.01]\n",
    "        sig_up_gs = list(sig_dge[sig_dge['logFC'] > 0].index.values)\n",
    "        sig_dn_gs = list(sig_dge[sig_dge['logFC'] < 0].index.values)\n",
    "        if len(sig_up_gs) > 2000:\n",
    "            sig_up_gs = sig_up_gs[:2000]\n",
    "        elif len(sig_up_gs) < 50:\n",
    "            sig_dge = dge[dge['adj.P.Val'] < 0.05]\n",
    "            sig_up_gs = list(sig_dge[sig_dge['logFC'] > 0].index.values)\n",
    "        if len(sig_dn_gs) > 2000:\n",
    "            sig_dn_gs = sig_dn_gs[:2000]\n",
    "        elif len(sig_dn_gs) < 50:\n",
    "            sig_dge = dge[dge['adj.P.Val'] < 0.05]\n",
    "            sig_dn_gs = list(sig_dge[sig_dge['logFC'] > 0].index.values)\n",
    "        cluster_dges[c]['up'] = sig_up_gs\n",
    "        cluster_dges[c]['down'] = sig_dn_gs\n",
    "        dge_table[c] = {}\n",
    "        dge_table[c]['up'] = len(cluster_dges[c]['up'])\n",
    "        dge_table[c]['down'] = len(cluster_dges[c]['down'])\n",
    "        display(FileLink(dge_filename, result_html_prefix=f'Differential expression for cluster {c}: '))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "\n",
    "if has_rna_file:    \n",
    "    targets = {}\n",
    "    bgs = {'GTEx': 'gtex-gene-stats.tsv', 'ARCHS4': 'archs4-gene-stats.tsv', 'TS': 'ts_10x_cell-ontology-class_donors_tissue-labels_v1.tsv'}\n",
    "    def find_targets(rna_df, bg, targets):\n",
    "        df_bg_stats = pd.read_csv(f\"https://appyters.maayanlab.cloud/storage/Tumor_Gene_Target_Screener/{bgs[bg]}\", sep='\\t', index_col=[0,1])\n",
    "        df_bg_genes = df_bg_stats.unstack().index.map(lambda idx: lookup(idx.partition('.')[0]))\n",
    "        df_bg_stats = df_bg_stats.unstack().groupby(df_bg_genes, observed=True).median().stack()\n",
    "        df_bg_expr = df_bg_stats.loc[(slice(None), ['25%', '50%', '75%']), :].unstack()\n",
    "        common_index = list(set(rna_df.index) & set(df_bg_expr.index))\n",
    "        expr_df = rna_df.loc[common_index, :]\n",
    "        index_name = expr_df.index.name\n",
    "        expr_df.reset_index(inplace=True)\n",
    "        expr_df.drop_duplicates(subset=index_name, inplace=True)\n",
    "        expr_df.set_index(index_name, inplace=True, drop=True)\n",
    "        target_distribution = df_bg_expr.loc[common_index, :].median(axis=1)\n",
    "        df_expr_norm = qnorm.quantile_normalize(expr_df.loc[common_index, :], target=target_distribution)\n",
    "        df_bg_expr_norm = qnorm.quantile_normalize(df_bg_expr.loc[common_index, :], target=target_distribution)\n",
    "        for cluster in tqdm(list(cluster_df['cluster'].unique())):\n",
    "            if cluster not in targets: targets[cluster] = {}\n",
    "            cluster_samples = list(cluster_df[cluster_df['cluster'] == cluster].index.values)\n",
    "            with suppress_output():\n",
    "                df_bg_expr_norm.columns = df_bg_expr_norm.columns.to_flat_index().map(lambda s: ', '.join(s))\n",
    "                dge = limma_voom_differential_expression(\n",
    "                    df_bg_expr_norm, df_expr_norm[cluster_samples],\n",
    "                    voom_design=True,\n",
    "                )\n",
    "                targets[cluster][bg] = dge[(dge['adj.P.Val'] < 0.01) & (dge['t'] > 0)].sort_values('t', ascending=False).index.values "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if has_rna_file: \n",
    "    print('Finding Targets using ARCHS4')\n",
    "    find_targets(rna_df, 'ARCHS4', targets)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if has_rna_file:    \n",
    "    print('Finding Targets using GTEx')\n",
    "    find_targets(rna_df, 'GTEx', targets)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if has_rna_file:      \n",
    "    print('Finding Targets using Tabula Sapiens')\n",
    "    find_targets(rna_df, 'TS', targets)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_eval\n",
    "\n",
    "if has_rna_file:\n",
    "    YlGnBu = cm.get_cmap('YlGnBu_r', 8)\n",
    "    cmap = {\"None\":YlGnBu(0), \"ARCHS4\": YlGnBu(1), \"GTEx\":YlGnBu(2), \"TS\": YlGnBu(4), \"ARCHS4-GTEx\":YlGnBu(3),  \"ARCHS4-TS\": YlGnBu(5), \"GTEx-TS\": YlGnBu(6), \"All\": YlGnBu(7)}\n",
    "\n",
    "    top_targets_n = 100\n",
    "    target_list = []\n",
    "    for cluster in targets:\n",
    "        for bg in targets[cluster]:\n",
    "            targets[cluster][bg] = list(filter(lambda x: not x.startswith('PCDH') ,targets[cluster][bg]))\n",
    "    for cluster in targets:\n",
    "        for bg in targets[cluster]:\n",
    "            target_list.extend(list(filter(lambda g: g in membrane_proteins, targets[cluster][bg][:top_targets_n])))\n",
    "    if len(set(target_list)) < 100:\n",
    "        top_targets_n = 500\n",
    "        for cluster in targets:\n",
    "            for bg in targets[cluster]:\n",
    "                target_list.extend(list(filter(lambda g: g in membrane_proteins, targets[cluster][bg][:top_targets_n])))\n",
    "                target_list.extend(list(targets[cluster][bg][:top_targets_n]))\n",
    "    \n",
    "    data1, data2, data3 = [], [], []\n",
    "    similarity = []\n",
    "    target_list = list(set(target_list))\n",
    "\n",
    "    for gene in target_list:\n",
    "        a = [1 if gene in targets[c]['ARCHS4'][:top_targets_n] else 0 for c in clusters]\n",
    "        g = [2 if gene in targets[c]['GTEx'][:top_targets_n] else 0 for c in clusters]\n",
    "        l = [4 if gene in targets[c]['TS'][:top_targets_n] else 0 for c in clusters]\n",
    "        data1.append(a)\n",
    "        data2.append(g)\n",
    "        data3.append(l)\n",
    "        similarity.append(np.dot(np.dot(np.array(a), np.array(g)), np.array(l)))\n",
    "\n",
    "    data = np.add(np.add(data1, data2), data3)\n",
    "\n",
    "    membrane_target_mat = pd.DataFrame(data)\n",
    "    membrane_target_mat.columns = [f\"Cluster {c}\" for c in clusters]\n",
    "    membrane_target_mat.index = target_list\n",
    "\n",
    "    membrane_target_mat['count'] = membrane_target_mat.sum(axis=1)\n",
    "    membrane_target_mat = membrane_target_mat[membrane_target_mat['count'] >= 7]\n",
    "    membrane_target_mat = membrane_target_mat.rename_axis('Membrane Target').sort_values(by = ['count', 'Membrane Target'], ascending = [False, True]).drop('count', axis=1)\n",
    "\n",
    "    h = membrane_target_mat.shape[0]\n",
    "\n",
    "    cluster_targets = {}\n",
    "    for col in membrane_target_mat.columns:\n",
    "        cluster_targets[col + ' Cell Surface Target'] = ','.join(list(membrane_target_mat[col][membrane_target_mat[col] >= 5].index.values))\n",
    "\n",
    "    g = sns.clustermap(membrane_target_mat, figsize=(4,0.3*h+2*(h<15)), cmap=YlGnBu, cbar_pos=None, dendrogram_ratio=0.1-(h<40)*0.01*(h-30), row_cluster=False, xticklabels=True, yticklabels=True)\n",
    "    g.ax_row_dendrogram.legend(handles=[Rectangle((0, 0), 0, 0, color=val, label=key) for key, val in cmap.items()],\n",
    "                                    title='Background', loc='upper right')\n",
    "\n",
    "    plt.show()\n",
    "\n",
    "    g.savefig(f'figures/{rna_filename}_membrane_target_matrix.png', dpi=300)\n",
    "    g.savefig(f'figures/{rna_filename}_membrane_target_matrix.svg', dpi=300)\n",
    "    membrane_target_mat.to_csv(f'results/{rna_filename}_membrane_target_mat.csv')\n",
    "    display(Markdown(f\"__Fig. {fig_counter}__ Top 100 consensus targets for each cluster with adjusted p-value < 0.01 compared to ARCHS4, GTEx, and Tabula Sapiens (TS) healthy backgrounds with limma voom.\"))\n",
    "    display(FileLink(f'results/{rna_filename}_membrane_target_mat.csv', result_html_prefix='Download Membrane Target Matrix: '))\n",
    "    display(FileLink(f'figures/{rna_filename}_membrane_target_matrix.png', result_html_prefix='Membrane Target Matrix PNG: '))\n",
    "    display(FileLink(f'figures/{rna_filename}_membrane_target_matrix.svg', result_html_prefix='Membrane Target Matrix SVG: '))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if has_rna_file:\n",
    "    display(Markdown(f\"Cell surface targets were identified for each cluster and are ordered by the number of healthy tissues and cell type background atlases that were used to identify the same target (Fig. {fig_counter}).\"))\n",
    "    fig_counter += 1"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
