{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "#%%appyter init\n",
    "from appyter import magic\n",
    "magic.init(lambda _=globals: _())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter hide_code\n",
    "\n",
    "{% do SectionField(\n",
    "    name='primary',\n",
    "    title='Upload Tumor Expression',\n",
    "    img='upload.png'\n",
    ") %}\n",
    "\n",
    "{% set data_type = TabField(\n",
    "    name='data_type',\n",
    "    label='Data Type',\n",
    "    default='scRNA-seq',\n",
    "    description='Start with either scRNA-seq or bulk RNA-seq data. If bulk RNA-seq data is selected we will utilize pre-curated reference matrices to identify cell type-specific expression vectors.',\n",
    "    required=True,\n",
    "    section='primary',\n",
    "    choices={\n",
    "        'scRNA-seq': [\n",
    "            TabField(\n",
    "            name='sc_data_type_ctrl',\n",
    "            label='Data Format (Control Profile)',\n",
    "            default='Plain Text',\n",
    "            description='Choose a format for scRNA-seq data',\n",
    "            required=True,\n",
    "            choices= {\n",
    "                'Plain Text': [\n",
    "                        CustomFileField(\n",
    "                        name='sc_rna_file_ctrl',\n",
    "                        label='control scRNA-seq',\n",
    "                        description='''\n",
    "    File should be a tsv/csv of the form:\n",
    "\n",
    "                                        cell 1   cell 2    ...\n",
    "                    ------------------------------\n",
    "            Gene/Protein 1    0          2       ...\n",
    "                    ------------------------------\n",
    "            Gene/Protein 2    1         1        ...\n",
    "                    ------------------------------\n",
    "                        ...                       ... \n",
    "                        ''',\n",
    "                        default='data/GSE171417_rbk_control.tsv.gz',\n",
    "                        required=False,\n",
    "                        examples={\n",
    "                            'GSE171417_rbk_control.tsv.gz': 'https://minio.dev.maayanlab.cloud/sc2targets/GSE171417_rbk_control.tsv.gz',\n",
    "                        },\n",
    "                    ),\n",
    "                    CustomFileField(\n",
    "                        name='sc_meta_file_ctrl',\n",
    "                        label='metadata',\n",
    "                        description='''\n",
    "                        File should be a tsv/csv of the form:\n",
    "\n",
    "                            Patient 1 Tumor    Patient 2 Tumor  ...\n",
    "                        ---------------------------------------\n",
    "                        Gene/Protein 1    0                   200       ...\n",
    "                        ---------------------------------------\n",
    "                        Gene/Protein 2    5                   180       ...\n",
    "                        ---------------------------------------\n",
    "                        ...                       ...                    ...        ...\n",
    "                        ''',\n",
    "                        default='data/GSE171417_metadata_control.tsv',\n",
    "                        required=False,\n",
    "                        examples={\n",
    "                            'GSE171417_metadata_control.tsv': 'https://minio.dev.maayanlab.cloud/sc2targets/GSE171417_metadata_control.tsv',\n",
    "                        },\n",
    "                    ),\n",
    "                    TextField(\n",
    "                        name='cell_type_col_ctrl',\n",
    "                        label='Cell Type Column (Optional)',\n",
    "                        description='Name of cell type column in uploaded metadata (if not selected, cell type identification will be performed)',\n",
    "                        default='',\n",
    "                        hint='cell_type',\n",
    "                        required=False\n",
    "                    )\n",
    "                    ],\n",
    "                '.mtx': [\n",
    "                        CustomFileField(\n",
    "                        name='sc_rna_file',\n",
    "                        label='scRNA-seq',\n",
    "                        description='''\n",
    "    File should be a tsv/csv of the form:\n",
    "\n",
    "                                        cell 1   cell 2    ...\n",
    "                    ------------------------------\n",
    "            Gene/Protein 1    0          2       ...\n",
    "                    ------------------------------\n",
    "            Gene/Protein 2    1         1        ...\n",
    "                    ------------------------------\n",
    "                        ...                       ... \n",
    "                        ''',\n",
    "                        default='data/sarcoma-v2-proteome-SpectrumMill-ratio-QCfilter-NArm.tsv',\n",
    "                        required=False,\n",
    "                        examples={\n",
    "                            'BR_proteomics_tumor.tsv': 'https://s3.amazonaws.com/multiomics2paper/public/BR_proteomics_tumor.tsv',\n",
    "                        },\n",
    "                    ),\n",
    "                    CustomFileField(\n",
    "                        name='sc_meta_file_pert',\n",
    "                        label='metadata',\n",
    "                        description='''\n",
    "                        File should be a tsv/csv of the form:\n",
    "\n",
    "                            Patient 1 Tumor    Patient 2 Tumor  ...\n",
    "                        ---------------------------------------\n",
    "                        Gene/Protein 1    0                   200       ...\n",
    "                        ---------------------------------------\n",
    "                        Gene/Protein 2    5                   180       ...\n",
    "                        ---------------------------------------\n",
    "                        ...                       ...                    ...        ...\n",
    "                        ''',\n",
    "                        default='data/sarcoma-v2-proteome-SpectrumMill-ratio-QCfilter-NArm.tsv',\n",
    "                        required=False,\n",
    "                        examples={\n",
    "                            'BR_proteomics_tumor.tsv': 'https://s3.amazonaws.com/multiomics2paper/public/BR_proteomics_tumor.tsv',\n",
    "                        },\n",
    "                    )]\n",
    "            }\n",
    "        ),\n",
    "        TabField(\n",
    "            name='sc_data_type',\n",
    "            label='Data format',\n",
    "            default='Plain Text',\n",
    "            description='Choose a format for scRNA-seq data',\n",
    "            required=True,\n",
    "            choices= {\n",
    "                'Plain Text': [\n",
    "                        CustomFileField(\n",
    "                        name='sc_rna_file_pert',\n",
    "                        label='scRNA-seq',\n",
    "                        description='''\n",
    "    File should be a tsv/csv of the form:\n",
    "\n",
    "                                        cell 1   cell 2    ...\n",
    "                    ------------------------------\n",
    "            Gene/Protein 1    0          2       ...\n",
    "                    ------------------------------\n",
    "            Gene/Protein 2    1         1        ...\n",
    "                    ------------------------------\n",
    "                        ...                       ... \n",
    "                        ''',\n",
    "                        default='data/GSE171417_rbk_experimental.tsv.gz',\n",
    "                        required=False,\n",
    "                        examples={\n",
    "                            'GSE171417_rbk_experimental.tsv.gz': 'https://minio.dev.maayanlab.cloud/sc2targets/GSE171417_rbk_experimental.tsv.gz',\n",
    "                        },\n",
    "                    ),\n",
    "                    CustomFileField(\n",
    "                        name='sc_meta_file_pert',\n",
    "                        label='metadata',\n",
    "                        description='''\n",
    "                        File should be a tsv/csv of the form:\n",
    "\n",
    "                            Patient 1 Tumor    Patient 2 Tumor  ...\n",
    "                        ---------------------------------------\n",
    "                        Gene/Protein 1    0                   200       ...\n",
    "                        ---------------------------------------\n",
    "                        Gene/Protein 2    5                   180       ...\n",
    "                        ---------------------------------------\n",
    "                        ...                       ...                    ...        ...\n",
    "                        ''',\n",
    "                        default='data/GSE171417_metadata_experimental.tsv',\n",
    "                        required=False,\n",
    "                        examples={\n",
    "                            'GSE171417_metadata_experimental.tsv': 'https://minio.dev.maayanlab.cloud/sc2targets/GSE171417_metadata_experimental.tsv',\n",
    "                        },\n",
    "                    ),\n",
    "                    TextField(\n",
    "                        name='condition_ctrl',\n",
    "                        label='Cell Type Column (Optional)',\n",
    "                        description='Name of the condition',\n",
    "                        default='',\n",
    "                        hint='cell_type',\n",
    "                        required=False\n",
    "                    ),\n",
    "                    ],\n",
    "                '.mtx': [\n",
    "                        CustomFileField(\n",
    "                        name='sc_rna_file',\n",
    "                        label='scRNA-seq',\n",
    "                        description='''\n",
    "    File should be a tsv/csv of the form:\n",
    "\n",
    "                                        cell 1   cell 2    ...\n",
    "                    ------------------------------\n",
    "            Gene/Protein 1    0          2       ...\n",
    "                    ------------------------------\n",
    "            Gene/Protein 2    1         1        ...\n",
    "                    ------------------------------\n",
    "                        ...                       ... \n",
    "                        ''',\n",
    "                        default='data/sarcoma-v2-proteome-SpectrumMill-ratio-QCfilter-NArm.tsv',\n",
    "                        required=False,\n",
    "                        examples={\n",
    "                            'BR_proteomics_tumor.tsv': 'https://s3.amazonaws.com/multiomics2paper/public/BR_proteomics_tumor.tsv',\n",
    "                        },\n",
    "                    ),\n",
    "                    CustomFileField(\n",
    "                        name='sc_meta_file',\n",
    "                        label='metadata',\n",
    "                        description='''\n",
    "                        File should be a tsv/csv of the form:\n",
    "\n",
    "                            Patient 1 Tumor    Patient 2 Tumor  ...\n",
    "                        ---------------------------------------\n",
    "                        Gene/Protein 1    0                   200       ...\n",
    "                        ---------------------------------------\n",
    "                        Gene/Protein 2    5                   180       ...\n",
    "                        ---------------------------------------\n",
    "                        ...                       ...                    ...        ...\n",
    "                        ''',\n",
    "                        default='data/sarcoma-v2-proteome-SpectrumMill-ratio-QCfilter-NArm.tsv',\n",
    "                        required=False,\n",
    "                        examples={\n",
    "                            'BR_proteomics_tumor.tsv': 'https://s3.amazonaws.com/multiomics2paper/public/BR_proteomics_tumor.tsv',\n",
    "                        },\n",
    "                    )]\n",
    "            }\n",
    "        )],\n",
    "        'bulk RNA-seq':[\n",
    "            CustomFileField(\n",
    "                name='bulk_expr',\n",
    "                label='RNA-seq expression',\n",
    "                description='''\n",
    "                File should be a tsv/csv of the form:\n",
    "\n",
    "                    Patient 1 Tumor    Patient 2 Tumor  ...\n",
    "                ---------------------------------------\n",
    "                Gene/Protein 1    0                   200       ...\n",
    "                ---------------------------------------\n",
    "                Gene/Protein 2    5                   180       ...\n",
    "                ---------------------------------------\n",
    "                ...                       ...                    ...        ...\n",
    "                ''',\n",
    "                default='data/GSE144441_Expression.tsv',\n",
    "                required=False,\n",
    "                examples={\n",
    "                    'GSE144441_Expression.tsv': 'https://minio.dev.maayanlab.cloud/sc2targets/GSE144441_Expression.tsv',\n",
    "                },\n",
    "            ),\n",
    "            CustomFileField(\n",
    "                        name='bulk_meta_file',\n",
    "                        label='metadata',\n",
    "                        description='''\n",
    "                        File should be a tsv/csv of the form:\n",
    "\n",
    "                            Patient 1 Tumor    Patient 2 Tumor  ...\n",
    "                        ---------------------------------------\n",
    "                        Gene/Protein 1    0                   200       ...\n",
    "                        ---------------------------------------\n",
    "                        Gene/Protein 2    5                   180       ...\n",
    "                        ---------------------------------------\n",
    "                        ...                       ...                    ...        ...\n",
    "                        ''',\n",
    "                        default='data/GSE144441_Metadata.tsv',\n",
    "                        required=False,\n",
    "                        examples={\n",
    "                            'GSE144441_Metadata.tsv': 'https://minio.dev.maayanlab.cloud/sc2targets/GSE144441_Metadata.tsv',\n",
    "                        },\n",
    "                    ),\n",
    "                    TextField(\n",
    "                        name='condition_col',\n",
    "                        label='Condition Column',\n",
    "                        description='Name of the condition column in uploaded metadata',\n",
    "                        default='Condition',\n",
    "                        required=True\n",
    "                    ),\n",
    "                    TextField(\n",
    "                        name='ctrl_condition',\n",
    "                        label='Control Condition Name',\n",
    "                        description='Name of the control condition in the condition column in uploaded metadata',\n",
    "                        default='healthy',\n",
    "                        required=True\n",
    "                    ),\n",
    "                    ChoiceField(\n",
    "                        name='reference',\n",
    "                        label='Single Cell Reference',\n",
    "                        description='Choose a single cell reference to use for cell type deconvolution',\n",
    "                        choices={\n",
    "                            \"Pancreas - Tabula Sapiens (Human)\": \"https://cellxgene.cziscience.com/collections/e5f58829-1a66-40b5-a624-9046778e74f5\",\n",
    "                            \"Pancreas - Tritschler et al. (Mouse)\": \"https://cellxgene.cziscience.com/collections/0a77d4c0-d5d0-40f0-aa1a-5e1429bcbd7e\",\n",
    "                            \"Adipose Tissue - Tabula Sapiens (Human)\": \"https://cellxgene.cziscience.com/collections/e5f58829-1a66-40b5-a624-9046778e74f5\",\n",
    "                            \"Adipose Tissue - Emont et al. (Mouse)\": \"https://cellxgene.cziscience.com/collections/fe0e718d-2ee9-42cc-894b-0b490f437dfd\",\n",
    "                            \"Adipose Tissue - Tabula Muris (Mouse)\": \"https://cellxgene.cziscience.com/collections/0b9d8a04-bb9d-44da-aa27-705bb65b54eb\",\n",
    "                            \"Liver - MacParland et al. 2018 (Human)\": \"https://cellxgene.cziscience.com/collections/bd5230f4-cd76-4d35-9ee5-89b3e7475659\",\n",
    "                            \"Blood - Tabula Sapiens (Human)\": \"https://cellxgene.cziscience.com/collections/e5f58829-1a66-40b5-a624-9046778e74f5\",\n",
    "                            \"Kidney - Tabula Sapiens(Human)\": \"https://cellxgene.cziscience.com/collections/e5f58829-1a66-40b5-a624-9046778e74f5\",\n",
    "                            \"Brain - Tabula Muris (Mouse)\": \"https://cellxgene.cziscience.com/collections/0b9d8a04-bb9d-44da-aa27-705bb65b54eb\",\n",
    "                            \"Skeletal Muscle - Domínguez Conde et al. 2022 (Human)\": \"https://cellxgene.cziscience.com/collections/62ef75e4-cbea-454e-a0ce-998ec40223d3\",\n",
    "                            \"Heart - Tabula Sapiens (Human)\": \"https://cellxgene.cziscience.com/collections/e5f58829-1a66-40b5-a624-9046778e74f5\",\n",
    "                            \"Heart - Litviňuková et al. 2020 (Mouse)\": \"https://cellxgene.cziscience.com/collections/b52eb423-5d0d-4645-b217-e1c6d38b2e72\",\n",
    "                            \"Heart - Tabula Muris (Mouse)\": \"https://cellxgene.cziscience.com/collections/0b9d8a04-bb9d-44da-aa27-705bb65b54eb\",\n",
    "                            \"Muscle - Tabula Sapiens (Human)\": \"https://cellxgene.cziscience.com/collections/e5f58829-1a66-40b5-a624-9046778e74f5\",\n",
    "                            \"Muscle - Tabula Muris (Mouse)\": \"https://cellxgene.cziscience.com/collections/0b9d8a04-bb9d-44da-aa27-705bb65b54eb\",\n",
    "                            \"Skin - Tabula Sapiens (Human)\": \"https://cellxgene.cziscience.com/collections/e5f58829-1a66-40b5-a624-9046778e74f5\",\n",
    "                            \"Retina - Li et al. 2023 (Human)\": \"https://cellxgene.cziscience.com/collections/4c6eaf5c-6d57-4c76-b1e9-60df8c655f1e\",\n",
    "                            \"Retina - Cowan et al. 2020 (Mouse)\": \"https://cellxgene.cziscience.com/collections/2f4c738f-e2f3-4553-9db2-0582a38ea4dc\",\n",
    "                            \"Breast - Reed et al. 2024 (Human)\": \"https://cellxgene.cziscience.com/collections/48259aa8-f168-4bf5-b797-af8e88da6637\",\n",
    "                            \"Lung - Tabula Sapiens (Human)\": \"https://cellxgene.cziscience.com/collections/e5f58829-1a66-40b5-a624-9046778e74f5\",\n",
    "                            \"Lung - Tabula Muris (Mouse)\": \"https://cellxgene.cziscience.com/collections/0b9d8a04-bb9d-44da-aa27-705bb65b54eb\",\n",
    "                            \"Uterus - Tabula Muris (Mouse)\": \"https://cellxgene.cziscience.com/collections/0b9d8a04-bb9d-44da-aa27-705bb65b54eb\",\n",
    "                            \"Bladder - Tabula Sapiens (Human)\": \"https://cellxgene.cziscience.com/collections/0b9d8a04-bb9d-44da-aa27-705bb65b54eb\",\n",
    "                            \"Bladder - Tabula Muris (Mouse)\": \"https://cellxgene.cziscience.com/collections/0b9d8a04-bb9d-44da-aa27-705bb65b54eb\",\n",
    "                            \"Gingiva - Easter et al. 2024 (Human)\": \"https://cellxgene.cziscience.com/collections/71f4bccf-53d4-4c12-9e80-e73bfb89e398\",\n",
    "                            \"Gingiva - Easter et al. 2024 (Mouse)\": \"https://cellxgene.cziscience.com/collections/67ba665e-0611-4b53-a522-40c2e0dc6df7\",\n",
    "                            \"Spleen - Tabula Muris (Mouse)\": \"https://cellxgene.cziscience.com/collections/0b9d8a04-bb9d-44da-aa27-705bb65b54eb\"\n",
    "                            }\n",
    "                    )\n",
    "         ]\n",
    "        }\n",
    ") %}\n",
    "\n",
    "\n",
    "{% set n_neighbors = IntField(\n",
    "    name='n_neighbors',\n",
    "    label='N neighbors', \n",
    "    default=15, \n",
    "    min=2,\n",
    "    max=100,\n",
    "    step=1,\n",
    "    description='The size of local neighborhood (in terms of number of neighboring data points) used for manifold approximation.', \n",
    "    section='primary'\n",
    ")\n",
    "%}\n",
    "\n",
    "{% set min_dist = FloatField(\n",
    "    name='min_dist',\n",
    "    label='Min Distance', \n",
    "    default=0.01, \n",
    "    min=0.00001,\n",
    "    max=0.9,\n",
    "    step=0.00001,\n",
    "    description='The effective minimum distance between embedded points.', \n",
    "    section='primary'\n",
    ")\n",
    "%}\n",
    "\n",
    "{% set resolution = FloatField(\n",
    "    name='resolution',\n",
    "    label='Resolution', \n",
    "    default=1, \n",
    "    min=0.1,\n",
    "    max=2,\n",
    "    step=0.1,\n",
    "    description='A parameter value controlling the coarseness of the clustering. Higher values lead to more clusters.', \n",
    "    section='primary'\n",
    ")\n",
    "%}\n",
    "\n",
    "{% set species = ChoiceField(\n",
    "    name='species',\n",
    "    label='Species', \n",
    "    default='mouse',\n",
    "    choices=['human', 'mouse'],\n",
    "    description='Choose a species', \n",
    "    required=True,\n",
    "    section='primary'\n",
    ")\n",
    "%}\n",
    "\n",
    "\n",
    "{% set membrane_screener_list = 'https://appyters.maayanlab.cloud/storage/Tumor_Gene_Target_Screener/surfaceome.csv' %}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [],
   "source": [
    "from helpers import *\n",
    "import os\n",
    "import re\n",
    "import random\n",
    "import qnorm\n",
    "from string import ascii_uppercase\n",
    "from tqdm import tqdm\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import decoupler as dc\n",
    "\n",
    "from matplotlib import cm\n",
    "from matplotlib.patches import Rectangle\n",
    "from matplotlib.patches import Patch\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from matplotlib.colors import LinearSegmentedColormap\n",
    "%matplotlib inline\n",
    "\n",
    "import networkx as nx\n",
    "\n",
    "from IPython.display import HTML, display, Markdown, FileLink\n",
    "from matplotlib.gridspec import GridSpec\n",
    "from matplotlib_venn import venn2\n",
    "from maayanlab_bioinformatics.normalization import zscore_normalize, log2_normalize\n",
    "from scipy.stats import zscore\n",
    "from scipy.stats import ttest_ind, ttest_1samp, ttest_ind_from_stats\n",
    "from maayanlab_bioinformatics.dge import limma_voom_differential_expression\n",
    "from maayanlab_bioinformatics.harmonization.ncbi_genes import ncbi_genes_lookup\n",
    "from maayanlab_bioinformatics.api import enrichr_link_from_genes\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "\n",
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "\n",
    "sc.settings.set_figure_params(dpi=300, frameon=False)\n",
    "sc.set_figure_params(dpi=300)\n",
    "sc.set_figure_params(figsize=(6, 6))\n",
    "\n",
    "os.makedirs('results', exist_ok=True)\n",
    "os.makedirs('figures', exist_ok=True)\n",
    "try:\n",
    "    lookup = ncbi_genes_lookup(organism='Mammalia/Homo_sapiens')\n",
    "except:\n",
    "    import urllib.request, json \n",
    "    with urllib.request.urlopen(\"https://s3.amazonaws.com/multiomics2paper/public/ncbi_genes_disambiguated.json\") as url:\n",
    "        ncbi_genes_disambiguated = json.load(url)\n",
    "        lookup = ncbi_genes_disambiguated.get\n",
    "    \n",
    "fig_counter = 2\n",
    "table_counter = 1\n",
    "letter_counter = 0\n",
    "discussion_results = {}\n",
    "\n",
    "import sys\n",
    "import contextlib\n",
    "@contextlib.contextmanager\n",
    "def suppress_output(stdout=True, stderr=True, dest=os.devnull):\n",
    "    ''' Usage:\n",
    "    with suppress_output():\n",
    "        print('hi')\n",
    "    '''\n",
    "    dev_null = open(dest, 'a')\n",
    "    if stdout:\n",
    "        _stdout = sys.stdout\n",
    "        sys.stdout = dev_null\n",
    "    if stderr:\n",
    "        _stderr = sys.stderr\n",
    "        sys.stderr = dev_null\n",
    "    try:\n",
    "        yield\n",
    "    finally:\n",
    "        if stdout:\n",
    "            sys.stdout = _stdout\n",
    "        if stderr:\n",
    "            sys.stderr = _stderr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "display(Markdown('## Abstract'))\n",
    "abstract_text = '''Single cell RNA-seq data enables the profiling of gene expression in individual cells. In disease states, cell type \n",
    "proportions and functions are often altered. By utilizing two single cell profiles from a control and experimental condition,\n",
    "TargetRanger can be used to identify highly expressed cell-surface proteins on diseased cell types that are lowly expressed on the healthy cell type and in healthy cells and tissues.\n",
    "Additionally, we can use L1000 data to identify compounds most likely to push diseased cell types to a healthy phenotype. SC2Targets also enables this analysis on bulk RNA-seq data, \n",
    "identifying cell-type specific expression for each bulk sample using a single cell reference and the the deconvolution algorithm, BayesPrism. Overall, SC2Targets enables the \n",
    "identification of cell-type specific targets that and reverser compounds to remove to treat disease. '''\n",
    "\n",
    "display(Markdown(abstract_text))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {
    "buttons": true
   },
   "outputs": [],
   "source": [
    "#load the buttons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_sc_data(sc_data_file: str, sc_metadata_file: str, type: str):\n",
    "    if type == 'plain':\n",
    "        if sc_data_file.endswith('.csv') or sc_data_file.endswith('.csz.gz'):\n",
    "            sc_data = pd.read_csv(sc_data_file, index_col=0, compression='gzip' if sc_data_file.endswith('.gz') else None)\n",
    "        elif sc_data_file.endswith('.txt') or sc_data_file.endswith('.txt.gz') or sc_data_file.endswith('.tsv') or sc_data_file.endswith('.tsv.gz'):\n",
    "            sc_data = pd.read_csv(sc_data_file, index_col=0, sep='\\t', compression='gzip' if sc_data_file.endswith('.gz') else None)\n",
    "        else:\n",
    "            raise ValueError('File type for scRNA-seq control profile not supported (.csv, .tsv, .txt)')\n",
    "        \n",
    "        if sc_metadata_file.endswith('.csv') or sc_metadata_file.endswith('.csz.gz'):\n",
    "            sc_metadata = pd.read_csv(sc_metadata_file, index_col=0, compression='gzip' if sc_metadata_file.endswith('.gz') else None)\n",
    "        elif sc_data_file.endswith('.txt') or sc_metadata_file.endswith('.txt.gz') or sc_data_file.endswith('.tsv') or sc_data_file.endswith('.tsv.gz'):\n",
    "            sc_metadata= pd.read_csv(sc_metadata_file, index_col=0, sep='\\t', compression='gzip' if sc_metadata_file.endswith('.gz') else None)\n",
    "        else:\n",
    "            raise ValueError('File type for scRNA-seq control profile not supported (.csv, .tsv, .txt)')\n",
    "        \n",
    "        adata = sc.AnnData(sc_data.T.values, obs=sc_metadata)\n",
    "        adata.var['gene_names'] = sc_data.index.values\n",
    "        adata.obs['samples'] = sc_metadata.index.values\n",
    "        return adata\n",
    "    \n",
    "def normalize_sc_data(adata, n_neighbors, min_dist, resolution):\n",
    "    sc.pp.filter_cells(adata, min_genes=200)\n",
    "    sc.pp.filter_genes(adata, min_cells=3)\n",
    "\n",
    "    # Annotate the group of mitochondrial genes as 'mt'\n",
    "    adata.var['mt'] = adata.var_names.str.startswith('MT-')\n",
    "    sc.pp.calculate_qc_metrics(adata, qc_vars=['mt'], percent_top=None, log1p=False, inplace=True)\n",
    "\n",
    "    # Filter cells following standard QC criteria.\n",
    "    adata = adata[adata.obs.n_genes_by_counts < 2500, :]\n",
    "    adata = adata[adata.obs.pct_counts_mt < 5, :]\n",
    "\n",
    "    # Normalize the data\n",
    "    sc.pp.normalize_total(adata, target_sum=1e4)\n",
    "    sc.pp.log1p(adata)\n",
    "    adata.layers['log_norm'] = adata.X.copy()\n",
    "    sc.pp.highly_variable_genes(adata, min_mean=0.0125, max_mean=3, min_disp=0.5)\n",
    "    sc.pp.regress_out(adata, ['total_counts', 'pct_counts_mt'])\n",
    "    sc.pp.scale(adata, max_value=10)\n",
    "    sc.tl.pca(adata, svd_solver='arpack')\n",
    "    dc.swap_layer(adata, 'log_norm', X_layer_key=None, inplace=True)\n",
    "    sc.pp.neighbors(adata, n_neighbors=n_neighbors, n_pcs=40)\n",
    "    sc.tl.umap(adata, min_dist=min_dist)\n",
    "    sc.tl.leiden(adata, resolution=resolution)\n",
    "    return adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "#load the data\n",
    "{% if data_type.raw_value == 'scRNA-seq' %}\n",
    "\n",
    "{% if data_type.value[0].raw_value == 'Plain Text' %}\n",
    "sc_data_ctrl_file = {{ data_type.value[0].value[0] }}\n",
    "sc_metadata_ctrl_file = {{ data_type.value[0].value[1] }}\n",
    "sc_data_ctrl = read_sc_data(sc_data_ctrl_file, sc_metadata_ctrl_file, 'plain')\n",
    "cell_type_col = {{ data_type.value[1].value[2] }}\n",
    "{% elif data_type.value[1].raw_value == '.mtx' %}\n",
    "## TODO IMPLEMENT .MTX FILE LOADING\n",
    "{% endif %}\n",
    "\n",
    "{% if data_type.value[1].raw_value == 'Plain Text' %}\n",
    "sc_data_pert_file = {{ data_type.value[1].value[0] }}\n",
    "sc_metadata_pert_file = {{ data_type.value[1].value[1] }}\n",
    "sc_data_pert = read_sc_data(sc_data_pert_file, sc_metadata_pert_file, 'plain')\n",
    "\n",
    "cell_type_col = {{ data_type.value[1].value[2] }}\n",
    "{% elif data_type.value[1].raw_value == '.mtx' %}\n",
    "## TODO IMPLEMENT .MTX FILE LOADING\n",
    "{% endif %}\n",
    "\n",
    "{% elif data_type.raw_value == 'bulk RNA-seq' %}\n",
    "## TODO READ BULK DATA, COMPUTE DECONVOLUTION\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "n_neighbors = {{ n_neighbors.raw_value }}\n",
    "min_dist = {{ min_dist.raw_value }}\n",
    "resolution = {{ resolution.raw_value }}\n",
    "species = '{{ species.raw_value }}'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc_data_pert: sc.AnnData = sc_data_pert\n",
    "sc_data_ctrl: sc.AnnData = sc_data_ctrl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% if data_type.raw_value == 'scRNA-seq' %}\n",
    "## Identify cell types if no cell type column is provided\n",
    "sc_data_ctrl = normalize_sc_data(sc_data_ctrl, n_neighbors, min_dist, resolution)\n",
    "sc.pl.umap(sc_data_ctrl, color='leiden', title='scRNA-seq control profile UMAP',\n",
    "           frameon=False, legend_fontweight='normal', legend_fontsize=15)\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% if data_type.raw_value == 'scRNA-seq' %}\n",
    "## Identify cell types if no cell type column is provided\n",
    "sc_data_pert = normalize_sc_data(sc_data_pert, n_neighbors, min_dist, resolution)\n",
    "sc.pl.umap(sc_data_ctrl, color='leiden', title='scRNA-seq perturbation profile UMAP',\n",
    "           frameon=False, legend_fontweight='normal', legend_fontsize=15)\n",
    "{% endif %}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "markers = pd.read_csv('https://minio.dev.maayanlab.cloud/sc2targets/PanglaoDB_markers_27_Mar_2020.tsv', sep='\\t')\n",
    "# Filter by canonical_marker and human\n",
    "markers['human'] = markers['species'].str.contains('Hs')\n",
    "markers['mouse'] = markers['species'].str.contains('Mm')\n",
    "markers = markers[markers[species] & (markers['canonical marker'] == 1) & (markers[f'sensitivity_{species}'] > 0.5)]\n",
    "markers = markers[~markers.duplicated(['cell type', 'official gene symbol'])]\n",
    "markers.rename(columns={'cell type': 'cell_type', 'official gene symbol': 'genesymbol'}, inplace=True)\n",
    "if species == 'mouse':\n",
    "    markers['genesymbol'] = markers['genesymbol'].map(lambda x: x.lower().capitalize())\n",
    "markers\n",
    "\n",
    "#sc_data_pert.var.set_index('gene_names', inplace=True)\n",
    "sc_data_ctrl.var.set_index('gene_names', inplace=True)\n",
    "\n",
    "dc.run_ora(\n",
    "    mat=sc_data_pert,\n",
    "    net=markers,\n",
    "    source='cell_type',\n",
    "    target='genesymbol',\n",
    "    min_n=3,\n",
    "    verbose=True,\n",
    "    use_raw=False\n",
    ")\n",
    "\n",
    "dc.run_ora(\n",
    "    mat=sc_data_ctrl,\n",
    "    net=markers,\n",
    "    source='cell_type',\n",
    "    target='genesymbol',\n",
    "    min_n=3,\n",
    "    verbose=True,\n",
    "    use_raw=False\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc_data_ctrl.obsm['ora_estimate']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 180,
   "metadata": {},
   "outputs": [],
   "source": [
    "acts_ctrl = dc.get_acts(sc_data_ctrl, obsm_key='ora_estimate')\n",
    "acts_pert = dc.get_acts(sc_data_pert, obsm_key='ora_estimate')\n",
    "sc_ctrl_cell_types = dc.rank_sources_groups(acts_ctrl, groupby='leiden', reference='rest', method='t-test_overestim_var')\n",
    "sc_pert_cell_types = dc.rank_sources_groups(acts_pert, groupby='leiden', reference='rest', method='t-test_overestim_var')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 183,
   "metadata": {},
   "outputs": [],
   "source": [
    "n_ctypes = 3\n",
    "ctypes_dict_ctrl = sc_ctrl_cell_types.groupby('group').head(n_ctypes).groupby('group')['names'].apply(lambda x: list(x)).to_dict()\n",
    "ctypes_dict_pert = sc_pert_cell_types.groupby('group').head(n_ctypes).groupby('group')['names'].apply(lambda x: list(x)).to_dict()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.matrixplot(acts_ctrl, ctypes_dict_ctrl, 'leiden', dendrogram=True, standard_scale='var',\n",
    "                 colorbar_title='Z-scaled scores', cmap='RdBu_r')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.matrixplot(acts_pert, ctypes_dict_pert, 'leiden', dendrogram=True, standard_scale='var',\n",
    "                 colorbar_title='Z-scaled scores', cmap='RdBu_r')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 186,
   "metadata": {},
   "outputs": [],
   "source": [
    "annotation_dict_ctrl = sc_ctrl_cell_types.groupby('group').head(1).set_index('group')['names'].to_dict()\n",
    "annotation_dict_pert = sc_pert_cell_types.groupby('group').head(1).set_index('group')['names'].to_dict()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc_data_ctrl.obs['cell_type'] = [annotation_dict_ctrl[clust] for clust in sc_data_ctrl.obs['leiden']]\n",
    "sc_data_pert.obs['cell_type'] = [annotation_dict_pert[clust] for clust in sc_data_pert.obs['leiden']]\n",
    "\n",
    "# Visualize\n",
    "sc.pl.umap(sc_data_ctrl, color='cell_type')\n",
    "sc.pl.umap(sc_data_pert, color='cell_type')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "overlapping_cell_types = list(set(annotation_dict_ctrl.values()).intersection(set(annotation_dict_pert.values())))\n",
    "overlapping_cell_types"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 219,
   "metadata": {},
   "outputs": [],
   "source": [
    "## get three representative samples per overlapping cell type\n",
    "data_pert = []\n",
    "sample_names_pert = []\n",
    "\n",
    "data_ctrl = []\n",
    "sample_names_ctrl = []\n",
    "# get indices of shared genes for each mat\n",
    "\n",
    "for ct in overlapping_cell_types:\n",
    "    ## extract counts from samples matching the cell type:\n",
    "    ct_cells = sc_data_pert.X.transpose()[:, sc_data_pert.obs['cell_type'] == ct]\n",
    "    n_cells = ct_cells.shape[1]\n",
    "    for i in range(3):\n",
    "        rand_idx = np.random.choice(list(range(n_cells)), n_cells // 3, replace=False)\n",
    "        ct_cells_rand_samp = ct_cells[:, rand_idx].sum(axis=1)\n",
    "        data_pert.append(ct_cells_rand_samp)\n",
    "        sample_names_pert.append(f'pert_{ct}_{i + 1}')\n",
    "    \n",
    "    ct_cells = sc_data_ctrl.X.transpose()[:, sc_data_ctrl.obs['cell_type'] == ct]\n",
    "    n_cells = ct_cells.shape[1]\n",
    "    for i in range(3):\n",
    "        rand_idx = np.random.choice(list(range(n_cells)), n_cells // 3, replace=False)\n",
    "        ct_cells_rand_samp = ct_cells[:, rand_idx].sum(axis=1)\n",
    "        data_ctrl.append(ct_cells_rand_samp)\n",
    "        sample_names_ctrl.append(f'ctrl_{ct}_{i + 1}')\n",
    "\n",
    "sim_bulk_pert = pd.DataFrame(data=data_pert, index=sample_names_pert, columns=sc_data_pert.var_names).T\n",
    "sim_bulk_ctrl = pd.DataFrame(data=data_ctrl, index=sample_names_ctrl, columns=sc_data_ctrl.var_names).T\n",
    "bulk_ct_df = pd.merge(sim_bulk_pert, sim_bulk_ctrl, left_index=True, right_index=True)\n",
    "bulk_ct_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    " \n",
    "targets = {}\n",
    "if species == 'mouse':\n",
    "    lookup = ncbi_genes_lookup(organism='Mammalia/Mus_musculus')\n",
    "    bgs = {'ARCHS4': 'Mammalia/Mus_musculus/archs4-gene-stats.tsv', 'Tabula Muris': 'Mammalia/Mus_musculus/tabula-muris-gene-stats.tsv' }\n",
    "else:\n",
    "    lookup = ncbi_genes_lookup()\n",
    "    bgs = {'GTEx': 'gtex-gene-stats.tsv', 'ARCHS4': 'archs4-gene-stats.tsv', 'TS': 'ts_10x_cell-ontology-class_donors_tissue-labels_v1.tsv'}\n",
    "def find_targets(rna_df, bg, targets):\n",
    "    df_bg_stats = pd.read_csv(f\"https://appyters.maayanlab.cloud/storage/Tumor_Gene_Target_Screener/{bgs[bg]}\", sep='\\t', index_col=[0,1])\n",
    "    df_bg_genes = df_bg_stats.unstack().index.map(lambda idx: lookup(idx.partition('.')[0]))\n",
    "    df_bg_stats = df_bg_stats.unstack().groupby(df_bg_genes, observed=True).median().stack()\n",
    "    df_bg_expr = df_bg_stats.loc[(slice(None), ['25%', '50%', '75%']), :].unstack()\n",
    "    common_index = list(set(rna_df.index) & set(df_bg_expr.index))\n",
    "    expr_df = rna_df.loc[common_index, :]\n",
    "    index_name = expr_df.index.name\n",
    "    expr_df.reset_index(inplace=True)\n",
    "    expr_df.drop_duplicates(subset=index_name, inplace=True)\n",
    "    expr_df.set_index(index_name, inplace=True, drop=True)\n",
    "    target_distribution = df_bg_expr.loc[common_index, :].median(axis=1)\n",
    "    df_expr_norm = qnorm.quantile_normalize(expr_df.loc[common_index, :], target=target_distribution)\n",
    "    df_bg_expr_norm = qnorm.quantile_normalize(df_bg_expr.loc[common_index, :], target=target_distribution)\n",
    "    for ct in tqdm(overlapping_cell_types):\n",
    "        if ct not in targets: targets[ct] = {}\n",
    "        ct_pert_samples = list(filter(lambda x: ct in x and 'pert' in x, bulk_ct_df.columns))\n",
    "        ct_ctrl_samples = list(filter(lambda x: ct in x and 'ctrl' in x, bulk_ct_df.columns))\n",
    "        with suppress_output():\n",
    "            df_bg_expr_norm.columns = df_bg_expr_norm.columns.to_flat_index().map(lambda s: ', '.join(s))\n",
    "            dge_pert = limma_voom_differential_expression(\n",
    "                df_bg_expr_norm, df_expr_norm[ct_pert_samples],\n",
    "                voom_design=True,\n",
    "            )\n",
    "            dge_ctrl = limma_voom_differential_expression(\n",
    "                df_bg_expr_norm, df_expr_norm[ct_ctrl_samples],\n",
    "                voom_design=True,\n",
    "            )\n",
    "            targets[ct][bg] = set(dge_pert[(dge_pert['adj.P.Val'] < 0.01) & (dge_pert['t'] > 0)].sort_values('t', ascending=False).index.values).difference(set(dge_ctrl[(dge_ctrl['adj.P.Val'] < 0.01) & (dge_ctrl['t'] > 0)].sort_values('t', ascending=False).index.values))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if species == 'human': \n",
    "    print('Finding Targets using ARCHS4')\n",
    "    find_targets(bulk_ct_df, 'ARCHS4', targets)  \n",
    "    print('Finding Targets using GTEx')\n",
    "    find_targets(bulk_ct_df, 'GTEx', targets)\n",
    "    print('Finding Targets using Tabula Sapiens')\n",
    "    find_targets(bulk_ct_df, 'TS', targets)\n",
    "elif species == 'mouse':\n",
    "    print('Finding Targets using ARCHS4')\n",
    "    find_targets(bulk_ct_df, 'ARCHS4', targets)\n",
    "    print('Finding Targets using Tabula Muris')\n",
    "    find_targets(bulk_ct_df, 'Tabula Muris', targets)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "proteins = pd.read_csv('https://appyters.maayanlab.cloud/storage/Tumor_Gene_Target_Screener/surfaceome.csv')\n",
    "membrane_proteins = list(proteins['genename'].map(lookup).dropna().values)\n",
    "\n",
    "if species == 'human':\n",
    "    YlGnBu = cm.get_cmap('YlGnBu_r', 8)\n",
    "    cmap = {\"None\":YlGnBu(0), \"ARCHS4\": YlGnBu(1), \"GTEx\":YlGnBu(2), \"TS\": YlGnBu(4), \"ARCHS4-GTEx\":YlGnBu(3),  \"ARCHS4-TS\": YlGnBu(5), \"GTEx-TS\": YlGnBu(6), \"All\": YlGnBu(7)}\n",
    "\n",
    "    target_list = []\n",
    "    for cluster in targets:\n",
    "        for bg in targets[cluster]:\n",
    "            targets[cluster][bg] = list(filter(lambda x: not x.startswith('PCDH') ,targets[cluster][bg]))\n",
    "    for cluster in targets:\n",
    "        for bg in targets[cluster]:\n",
    "            target_list.extend(list(filter(lambda g: g in membrane_proteins, targets[cluster][bg][:top_targets_n])))\n",
    "    if len(set(target_list)) < 100:\n",
    "        top_targets_n = 500\n",
    "        for cluster in targets:\n",
    "            for bg in targets[cluster]:\n",
    "                target_list.extend(list(filter(lambda g: g in membrane_proteins, targets[cluster][bg][:top_targets_n])))\n",
    "                target_list.extend(list(targets[cluster][bg][:top_targets_n]))\n",
    "    \n",
    "    data1, data2, data3 = [], [], []\n",
    "    similarity = []\n",
    "    target_list = list(set(target_list))\n",
    "\n",
    "    for gene in target_list:\n",
    "        a = [1 if gene in targets[c]['ARCHS4'][:top_targets_n] else 0 for c in clusters]\n",
    "        g = [2 if gene in targets[c]['GTEx'][:top_targets_n] else 0 for c in clusters]\n",
    "        l = [4 if gene in targets[c]['TS'][:top_targets_n] else 0 for c in clusters]\n",
    "        data1.append(a)\n",
    "        data2.append(g)\n",
    "        data3.append(l)\n",
    "        similarity.append(np.dot(np.dot(np.array(a), np.array(g)), np.array(l)))\n",
    "\n",
    "    data = np.add(np.add(data1, data2), data3)\n",
    "\n",
    "    membrane_target_mat = pd.DataFrame(data)\n",
    "    membrane_target_mat.columns = [f\"Cluster {c}\" for c in clusters]\n",
    "    membrane_target_mat.index = target_list\n",
    "\n",
    "    membrane_target_mat['count'] = membrane_target_mat.sum(axis=1)\n",
    "    membrane_target_mat = membrane_target_mat[membrane_target_mat['count'] >= 7]\n",
    "    membrane_target_mat = membrane_target_mat.rename_axis('Membrane Target').sort_values(by = ['count', 'Membrane Target'], ascending = [False, True]).drop('count', axis=1)\n",
    "\n",
    "    h = membrane_target_mat.shape[0]\n",
    "\n",
    "    cluster_targets = {}\n",
    "    for col in membrane_target_mat.columns:\n",
    "        cluster_targets[col + ' Cell Surface Target'] = ','.join(list(membrane_target_mat[col][membrane_target_mat[col] >= 5].index.values))\n",
    "\n",
    "    g = sns.clustermap(membrane_target_mat, figsize=(4,0.3*h+2*(h<15)), cmap=YlGnBu, cbar_pos=None, dendrogram_ratio=0.1-(h<40)*0.01*(h-30), row_cluster=False, xticklabels=True, yticklabels=True)\n",
    "    g.ax_row_dendrogram.legend(handles=[Rectangle((0, 0), 0, 0, color=val, label=key) for key, val in cmap.items()],\n",
    "                                    title='Background', loc='upper right')\n",
    "\n",
    "    plt.show()\n",
    "elif species == 'mouse':\n",
    "\n",
    "    YlGnBu = cm.get_cmap('YlGnBu_r', 4)\n",
    "    cmap = {\"None\": YlGnBu(0), \"ARCHS4\": YlGnBu(1), \"Tabula Muris\":YlGnBu(2), \"Both\": YlGnBu(3)}\n",
    "\n",
    "    top_targets_n = 100\n",
    "    target_list = []\n",
    "    for ct in targets:\n",
    "        for bg in targets[ct]:\n",
    "            #target_list.extend(list(filter(lambda g: g.upper() in membrane_proteins, list(targets[ct][bg]))))\n",
    "            target_list.extend(list(targets[ct][bg])[:20])\n",
    "\n",
    "    data1, data2 = [], []\n",
    "    similarity = []\n",
    "    target_list = list(set(target_list))\n",
    "\n",
    "    for gene in target_list:\n",
    "        a = [1 if gene in targets[c]['ARCHS4'] else 0 for c in overlapping_cell_types]\n",
    "        g = [2 if gene in targets[c]['Tabula Muris'] else 0 for c in overlapping_cell_types]\n",
    "        data1.append(a)\n",
    "        data2.append(g)\n",
    "        similarity.append(np.dot(np.array(a), np.array(g)))\n",
    "\n",
    "    data = np.add(data1, data2)\n",
    "\n",
    "    membrane_target_mat = pd.DataFrame(data)\n",
    "    membrane_target_mat.columns = overlapping_cell_types\n",
    "    membrane_target_mat.index = target_list\n",
    "    \n",
    "    membrane_target_mat['count'] = membrane_target_mat.sum(axis=1)\n",
    "    membrane_target_mat = membrane_target_mat[membrane_target_mat['count'] >= 3]\n",
    "    \n",
    "    membrane_target_mat = membrane_target_mat.rename_axis('Membrane Target').sort_values(by = ['count', 'Membrane Target'], ascending = [False, True]).drop('count', axis=1)\n",
    "\n",
    "    h = membrane_target_mat.shape[0]\n",
    "    membrane_target_mat = membrane_target_mat.astype(float)\n",
    "    \n",
    "    g = sns.clustermap(membrane_target_mat, figsize=(4,0.3*h+2*(h<15)), cmap=YlGnBu, cbar_pos=None, dendrogram_ratio=0.1-(h<40)*0.01*(h-30), row_cluster=False, col_cluster=False, xticklabels=True, yticklabels=True)\n",
    "    g.ax_row_dendrogram.legend(handles=[Rectangle((0, 0), 0, 0, color=val, label=key) for key, val in cmap.items()],\n",
    "                                    title='Background', loc='upper right')\n",
    "    plt.grid(False)\n",
    "\n",
    "\n",
    "    plt.show()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
