{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "#%%appyter init\n",
    "from appyter import magic\n",
    "magic.init(lambda _=globals: _())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter hide_code\n",
    "\n",
    "{% do SectionField(\n",
    "    name='primary',\n",
    "    title='Upload Tumor Expression',\n",
    "    img='upload.png'\n",
    ") %}\n",
    "\n",
    "{% set data_type = TabField(\n",
    "    name='data_type',\n",
    "    label='Data Type',\n",
    "    default='scRNA-seq',\n",
    "    description='Start with either scRNA-seq or bulk RNA-seq data. If bulk RNA-seq data is selected we will utilize pre-curated reference matrices to identify cell type-specific expression vectors.',\n",
    "    required=True,\n",
    "    section='primary',\n",
    "    choices={\n",
    "        'scRNA-seq': [\n",
    "            TabField(\n",
    "            name='sc_data_type_ctrl',\n",
    "            label='Data Format (Control Profile)',\n",
    "            default='Plain Text',\n",
    "            description='Choose a format for scRNA-seq data',\n",
    "            required=True,\n",
    "            choices= {\n",
    "                'Plain Text': [\n",
    "                        CustomFileField(\n",
    "                        name='sc_rna_file_ctrl',\n",
    "                        label='control scRNA-seq',\n",
    "                        description='''\n",
    "    File should be a tsv/csv of the form:\n",
    "\n",
    "                                        cell 1   cell 2    ...\n",
    "                    ------------------------------\n",
    "            Gene/Protein 1    0          2       ...\n",
    "                    ------------------------------\n",
    "            Gene/Protein 2    1         1        ...\n",
    "                    ------------------------------\n",
    "                        ...                       ... \n",
    "                        ''',\n",
    "                        default='data/GSE171417_rbk_control.tsv.gz',\n",
    "                        required=False,\n",
    "                        examples={\n",
    "                            'GSE171417_rbk_control.tsv.gz': 'https://minio.dev.maayanlab.cloud/sc2targets/GSE171417_rbk_control.tsv.gz',\n",
    "                        },\n",
    "                    ),\n",
    "                    CustomFileField(\n",
    "                        name='sc_meta_file_ctrl',\n",
    "                        label='metadata',\n",
    "                        description='''\n",
    "                        File should be a tsv/csv of the form:\n",
    "\n",
    "                            Patient 1 Tumor    Patient 2 Tumor  ...\n",
    "                        ---------------------------------------\n",
    "                        Gene/Protein 1    0                   200       ...\n",
    "                        ---------------------------------------\n",
    "                        Gene/Protein 2    5                   180       ...\n",
    "                        ---------------------------------------\n",
    "                        ...                       ...                    ...        ...\n",
    "                        ''',\n",
    "                        default='data/GSE171417_metadata_control.tsv',\n",
    "                        required=False,\n",
    "                        examples={\n",
    "                            'GSE171417_metadata_control.tsv': 'https://minio.dev.maayanlab.cloud/sc2targets/GSE171417_metadata_control.tsv',\n",
    "                        },\n",
    "                    ),\n",
    "                    TextField(\n",
    "                        name='cell_type_col_ctrl',\n",
    "                        label='Cell Type Column (Optional)',\n",
    "                        description='Name of cell type column in uploaded metadata (if not selected, cell type identification will be performed)',\n",
    "                        default='',\n",
    "                        hint='cell_type',\n",
    "                        required=False\n",
    "                    )\n",
    "                    ],\n",
    "                '.mtx': [\n",
    "                        CustomFileField(\n",
    "                        name='sc_rna_file',\n",
    "                        label='scRNA-seq',\n",
    "                        description='''\n",
    "    File should be a tsv/csv of the form:\n",
    "\n",
    "                                        cell 1   cell 2    ...\n",
    "                    ------------------------------\n",
    "            Gene/Protein 1    0          2       ...\n",
    "                    ------------------------------\n",
    "            Gene/Protein 2    1         1        ...\n",
    "                    ------------------------------\n",
    "                        ...                       ... \n",
    "                        ''',\n",
    "                        default='data/sarcoma-v2-proteome-SpectrumMill-ratio-QCfilter-NArm.tsv',\n",
    "                        required=False,\n",
    "                        examples={\n",
    "                            'BR_proteomics_tumor.tsv': 'https://s3.amazonaws.com/multiomics2paper/public/BR_proteomics_tumor.tsv',\n",
    "                        },\n",
    "                    ),\n",
    "                    CustomFileField(\n",
    "                        name='sc_meta_file_pert',\n",
    "                        label='metadata',\n",
    "                        description='''\n",
    "                        File should be a tsv/csv of the form:\n",
    "\n",
    "                            Patient 1 Tumor    Patient 2 Tumor  ...\n",
    "                        ---------------------------------------\n",
    "                        Gene/Protein 1    0                   200       ...\n",
    "                        ---------------------------------------\n",
    "                        Gene/Protein 2    5                   180       ...\n",
    "                        ---------------------------------------\n",
    "                        ...                       ...                    ...        ...\n",
    "                        ''',\n",
    "                        default='data/sarcoma-v2-proteome-SpectrumMill-ratio-QCfilter-NArm.tsv',\n",
    "                        required=False,\n",
    "                        examples={\n",
    "                            'BR_proteomics_tumor.tsv': 'https://s3.amazonaws.com/multiomics2paper/public/BR_proteomics_tumor.tsv',\n",
    "                        },\n",
    "                    )]\n",
    "            }\n",
    "        ),\n",
    "        TabField(\n",
    "            name='sc_data_type',\n",
    "            label='Data format',\n",
    "            default='Plain Text',\n",
    "            description='Choose a format for scRNA-seq data',\n",
    "            required=True,\n",
    "            choices= {\n",
    "                'Plain Text': [\n",
    "                        CustomFileField(\n",
    "                        name='sc_rna_file_pert',\n",
    "                        label='scRNA-seq',\n",
    "                        description='''\n",
    "    File should be a tsv/csv of the form:\n",
    "\n",
    "                                        cell 1   cell 2    ...\n",
    "                    ------------------------------\n",
    "            Gene/Protein 1    0          2       ...\n",
    "                    ------------------------------\n",
    "            Gene/Protein 2    1         1        ...\n",
    "                    ------------------------------\n",
    "                        ...                       ... \n",
    "                        ''',\n",
    "                        default='data/GSE171417_rbk_experimental.tsv.gz',\n",
    "                        required=False,\n",
    "                        examples={\n",
    "                            'GSE171417_rbk_experimental.tsv.gz': 'https://minio.dev.maayanlab.cloud/sc2targets/GSE171417_rbk_experimental.tsv.gz',\n",
    "                        },\n",
    "                    ),\n",
    "                    CustomFileField(\n",
    "                        name='sc_meta_file_pert',\n",
    "                        label='metadata',\n",
    "                        description='''\n",
    "                        File should be a tsv/csv of the form:\n",
    "\n",
    "                            Patient 1 Tumor    Patient 2 Tumor  ...\n",
    "                        ---------------------------------------\n",
    "                        Gene/Protein 1    0                   200       ...\n",
    "                        ---------------------------------------\n",
    "                        Gene/Protein 2    5                   180       ...\n",
    "                        ---------------------------------------\n",
    "                        ...                       ...                    ...        ...\n",
    "                        ''',\n",
    "                        default='data/GSE171417_metadata_experimental.tsv',\n",
    "                        required=False,\n",
    "                        examples={\n",
    "                            'GSE171417_metadata_experimental.tsv': 'https://minio.dev.maayanlab.cloud/sc2targets/GSE171417_metadata_experimental.tsv',\n",
    "                        },\n",
    "                    ),\n",
    "                    TextField(\n",
    "                        name='condition_ctrl',\n",
    "                        label='Cell Type Column (Optional)',\n",
    "                        description='Name of the condition',\n",
    "                        default='',\n",
    "                        hint='cell_type',\n",
    "                        required=False\n",
    "                    ),\n",
    "                    ],\n",
    "                '.mtx': [\n",
    "                        CustomFileField(\n",
    "                        name='sc_rna_file',\n",
    "                        label='scRNA-seq',\n",
    "                        description='''\n",
    "    File should be a tsv/csv of the form:\n",
    "\n",
    "                                        cell 1   cell 2    ...\n",
    "                    ------------------------------\n",
    "            Gene/Protein 1    0          2       ...\n",
    "                    ------------------------------\n",
    "            Gene/Protein 2    1         1        ...\n",
    "                    ------------------------------\n",
    "                        ...                       ... \n",
    "                        ''',\n",
    "                        default='data/sarcoma-v2-proteome-SpectrumMill-ratio-QCfilter-NArm.tsv',\n",
    "                        required=False,\n",
    "                        examples={\n",
    "                            'BR_proteomics_tumor.tsv': 'https://s3.amazonaws.com/multiomics2paper/public/BR_proteomics_tumor.tsv',\n",
    "                        },\n",
    "                    ),\n",
    "                    CustomFileField(\n",
    "                        name='sc_meta_file',\n",
    "                        label='metadata',\n",
    "                        description='''\n",
    "                        File should be a tsv/csv of the form:\n",
    "\n",
    "                            Patient 1 Tumor    Patient 2 Tumor  ...\n",
    "                        ---------------------------------------\n",
    "                        Gene/Protein 1    0                   200       ...\n",
    "                        ---------------------------------------\n",
    "                        Gene/Protein 2    5                   180       ...\n",
    "                        ---------------------------------------\n",
    "                        ...                       ...                    ...        ...\n",
    "                        ''',\n",
    "                        default='data/sarcoma-v2-proteome-SpectrumMill-ratio-QCfilter-NArm.tsv',\n",
    "                        required=False,\n",
    "                        examples={\n",
    "                            'BR_proteomics_tumor.tsv': 'https://s3.amazonaws.com/multiomics2paper/public/BR_proteomics_tumor.tsv',\n",
    "                        },\n",
    "                    )]\n",
    "            }\n",
    "        )],\n",
    "        'bulk RNA-seq':[\n",
    "            CustomFileField(\n",
    "                name='bulk_expr',\n",
    "                label='RNA-seq expression',\n",
    "                description='''\n",
    "                File should be a tsv/csv of the form:\n",
    "\n",
    "                    Patient 1 Tumor    Patient 2 Tumor  ...\n",
    "                ---------------------------------------\n",
    "                Gene/Protein 1    0                   200       ...\n",
    "                ---------------------------------------\n",
    "                Gene/Protein 2    5                   180       ...\n",
    "                ---------------------------------------\n",
    "                ...                       ...                    ...        ...\n",
    "                ''',\n",
    "                default='data/GSE49155_expression.tsv',\n",
    "                required=False,\n",
    "                examples={\n",
    "                    'GSE49155_expression.tsv': 'https://minio.dev.maayanlab.cloud/sc2targets/GSE49155_expression.tsv',\n",
    "                },\n",
    "            ),\n",
    "            CustomFileField(\n",
    "                        name='bulk_meta_file',\n",
    "                        label='metadata',\n",
    "                        description='''\n",
    "                        File should be a tsv/csv of the form:\n",
    "\n",
    "                            Patient 1 Tumor    Patient 2 Tumor  ...\n",
    "                        ---------------------------------------\n",
    "                        Gene/Protein 1    0                   200       ...\n",
    "                        ---------------------------------------\n",
    "                        Gene/Protein 2    5                   180       ...\n",
    "                        ---------------------------------------\n",
    "                        ...                       ...                    ...        ...\n",
    "                        ''',\n",
    "                        default='data/GSE49155_metadata.tsv',\n",
    "                        required=False,\n",
    "                        examples={\n",
    "                            'GSE49155_metadata.tsv': 'https://minio.dev.maayanlab.cloud/sc2targets/GSE49155_metadata.tsv',\n",
    "                        },\n",
    "                    ),\n",
    "                    TextField(\n",
    "                        name='condition_col',\n",
    "                        label='Condition Column',\n",
    "                        description='Name of the condition column in uploaded metadata',\n",
    "                        default='condition',\n",
    "                        required=True\n",
    "                    ),\n",
    "                    TextField(\n",
    "                        name='ctrl_condition',\n",
    "                        label='Control Condition Name',\n",
    "                        description='Name of the control condition in the condition column in uploaded metadata',\n",
    "                        default='normal',\n",
    "                        required=True\n",
    "                    ),\n",
    "                    ChoiceField(\n",
    "                        name='reference',\n",
    "                        label='Single Cell Reference',\n",
    "                        description='Choose a single cell reference to use for cell type deconvolution',\n",
    "                        default='Kidney - Tabula Sapiens (Human)',\n",
    "                        required=True,\n",
    "                        choices={\n",
    "                            \"Pancreas - Tabula Sapiens (Human)\": \"tabula_sapiens_pancreas\",\n",
    "                            \"Adipose Tissue - Tabula Sapiens (Human)\": \"tabula_sapiens_adipose\",\n",
    "                            \"Adipose Tissue - Tabula Muris (Mouse)\": \"tabula_muris_adipose\",\n",
    "                            \"Liver - Tabula Sapiens (Human)\": \"tabula_sapiens_liver\",\n",
    "                            \"Blood - Tabula Sapiens (Human)\": \"tabula_sapiens_blood\",\n",
    "                            \"Kidney - Tabula Sapiens (Human)\": \"tabula_sapiens_kidney\",\n",
    "                            \"Kidney - Lake et al. 2023 (Human)\": \"human_kidney_lake_et_al_2023\",\n",
    "                            \"Brain (non-myeloid) - Tabula Muris (Mouse)\": \"tabula_muris_brain-non-myeloid\",\n",
    "                            \"Heart - Tabula Sapiens (Human)\": \"tabula_sapiens_heart\",\n",
    "                            \"Heart - Tabula Muris (Mouse)\": \"tabula_muris_heart\",\n",
    "                            \"Muscle - Tabula Sapiens (Human)\": \"tabula_sapiens_muscle\",\n",
    "                            \"Muscle - Tabula Muris (Mouse)\": \"tabula_muris_muscle\",\n",
    "                            \"Skin - Tabula Sapiens (Human)\": \"tabula_sapiens_skin\",\n",
    "                            \"Lung - Tabula Sapiens (Human)\": \"tabula_sapiens_lung\",\n",
    "                            \"Lung - Tabula Muris (Mouse)\": \"tabula_muris_lung\",\n",
    "                            \"Spleen - Tabula Sapiens (Human)\": \"tabula_sapiens_spleen\",\n",
    "                            \"Spleen - Tabula Muris (Mouse)\": \"tabula_muris_spleen\",\n",
    "                            \"Prostate - Tabula Sapiens (Human)\": \"tabula_sapiens_prostate\",\n",
    "                            \"Lung Cancer Atlas (LuCA) - Salcher et al. 2022 (Human)\": \"salcher-et-al-lung-cancer\",\n",
    "                            }\n",
    "                    )\n",
    "         ]\n",
    "        }\n",
    ") %}\n",
    "\n",
    "#\"Pancreas - Tritschler et al. (Mouse)\": \"https://cellxgene.cziscience.com/collections/0a77d4c0-d5d0-40f0-aa1a-5e1429bcbd7e\",\n",
    "#\"Adipose Tissue - Emont et al. (Mouse)\": \"https://cellxgene.cziscience.com/collections/fe0e718d-2ee9-42cc-894b-0b490f437dfd\",\n",
    "#\"Liver - MacParland et al. 2018 (Human)\": \"https://cellxgene.cziscience.com/collections/bd5230f4-cd76-4d35-9ee5-89b3e7475659\",\n",
    "#\"Skeletal Muscle - Domínguez Conde et al. 2022 (Human)\": \"https://cellxgene.cziscience.com/collections/62ef75e4-cbea-454e-a0ce-998ec40223d3\",\n",
    "#\"Heart - Litviňuková et al. 2020 (Mouse)\": \"https://cellxgene.cziscience.com/collections/b52eb423-5d0d-4645-b217-e1c6d38b2e72\",\n",
    "#\"Retina - Li et al. 2023 (Human)\": \"https://cellxgene.cziscience.com/collections/4c6eaf5c-6d57-4c76-b1e9-60df8c655f1e\",\n",
    "#\"Retina - Cowan et al. 2020 (Mouse)\": \"https://cellxgene.cziscience.com/collections/2f4c738f-e2f3-4553-9db2-0582a38ea4dc\",\n",
    "#Breast - Reed et al. 2024 (Human)\": \"https://cellxgene.cziscience.com/collections/48259aa8-f168-4bf5-b797-af8e88da6637\",\n",
    "\n",
    "\n",
    "{% set n_neighbors = IntField(\n",
    "    name='n_neighbors',\n",
    "    label='N neighbors', \n",
    "    default=15, \n",
    "    min=2,\n",
    "    max=100,\n",
    "    step=1,\n",
    "    description='The size of local neighborhood (in terms of number of neighboring data points) used for manifold approximation.', \n",
    "    section='primary'\n",
    ")\n",
    "%}\n",
    "\n",
    "{% set min_dist = FloatField(\n",
    "    name='min_dist',\n",
    "    label='Min Distance', \n",
    "    default=0.01, \n",
    "    min=0.00001,\n",
    "    max=0.9,\n",
    "    step=0.00001,\n",
    "    description='The effective minimum distance between embedded points.', \n",
    "    section='primary'\n",
    ")\n",
    "%}\n",
    "\n",
    "{% set resolution = FloatField(\n",
    "    name='resolution',\n",
    "    label='Resolution', \n",
    "    default=1, \n",
    "    min=0.1,\n",
    "    max=2,\n",
    "    step=0.1,\n",
    "    description='A parameter value controlling the coarseness of the clustering. Higher values lead to more clusters.', \n",
    "    section='primary'\n",
    ")\n",
    "%}\n",
    "\n",
    "{% set species = ChoiceField(\n",
    "    name='species',\n",
    "    label='Species', \n",
    "    default='mouse',\n",
    "    choices=['human', 'mouse'],\n",
    "    description='Choose a species', \n",
    "    required=True,\n",
    "    section='primary'\n",
    ")\n",
    "%}\n",
    "\n",
    "\n",
    "{% set membrane_screener_list = 'https://appyters.maayanlab.cloud/storage/Tumor_Gene_Target_Screener/surfaceome.csv' %}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "from helpers import *\n",
    "import os\n",
    "import re\n",
    "import random\n",
    "import qnorm\n",
    "import shutil\n",
    "from string import ascii_uppercase\n",
    "from tqdm import tqdm\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import decoupler as dc\n",
    "\n",
    "from matplotlib import cm\n",
    "from matplotlib.patches import Rectangle\n",
    "from matplotlib.patches import Patch\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from matplotlib.colors import LinearSegmentedColormap\n",
    "%matplotlib inline\n",
    "\n",
    "import networkx as nx\n",
    "\n",
    "from IPython.display import HTML, display, Markdown, FileLink\n",
    "from matplotlib.gridspec import GridSpec\n",
    "from matplotlib_venn import venn2\n",
    "from maayanlab_bioinformatics.normalization import zscore_normalize, log2_normalize\n",
    "from maayanlab_bioinformatics.dge import ttest_differential_expression\n",
    "from maayanlab_bioinformatics.harmonization.ncbi_genes import ncbi_genes_lookup\n",
    "from maayanlab_bioinformatics.api import enrichr_link_from_genes\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "\n",
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "\n",
    "sc.settings.set_figure_params(dpi=300, frameon=False)\n",
    "sc.settings.verbosity = 0\n",
    "\n",
    "os.makedirs('results', exist_ok=True)\n",
    "os.makedirs('figures', exist_ok=True)\n",
    "\n",
    "    \n",
    "fig_counter = 2\n",
    "table_counter = 1\n",
    "letter_counter = 0\n",
    "discussion_results = {}\n",
    "\n",
    "import sys\n",
    "import contextlib\n",
    "@contextlib.contextmanager\n",
    "def suppress_output(stdout=True, stderr=True, dest=os.devnull):\n",
    "    ''' Usage:\n",
    "    with suppress_output():\n",
    "        print('hi')\n",
    "    '''\n",
    "    dev_null = open(dest, 'a')\n",
    "    if stdout:\n",
    "        _stdout = sys.stdout\n",
    "        sys.stdout = dev_null\n",
    "    if stderr:\n",
    "        _stderr = sys.stderr\n",
    "        sys.stderr = dev_null\n",
    "    try:\n",
    "        yield\n",
    "    finally:\n",
    "        if stdout:\n",
    "            sys.stdout = _stdout\n",
    "        if stderr:\n",
    "            sys.stderr = _stderr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "display(Markdown('## Abstract'))\n",
    "abstract_text = '''Single cell RNA-seq data enables the profiling of gene expression in individual cells. In disease states, cell type \n",
    "proportions and functions are often altered. By utilizing two single cell profiles from a control and experimental condition,\n",
    "TargetRanger can be used to identify highly expressed cell-surface proteins on diseased cell types that are lowly expressed on the healthy cell type and in healthy cells and tissues.\n",
    "Additionally, we can use L1000 data to identify compounds most likely to push diseased cell types to a healthy phenotype. SC2Targets also enables this analysis on bulk RNA-seq data, \n",
    "identifying cell-type specific expression for each bulk sample using a single cell reference and the the deconvolution algorithm, BayesPrism. Overall, SC2Targets enables the \n",
    "identification of cell-type specific targets that and reverser compounds to remove to treat disease. '''\n",
    "\n",
    "display(Markdown(abstract_text))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "buttons": true
   },
   "outputs": [],
   "source": [
    "#load the buttons"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Methods"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "## TODO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "<img src='https://minio.dev.maayanlab.cloud/sc2targets/scRNAseq2TargetsWorkflow.png' alt=\"scRNAseq2Targets Workflow Diagram\">"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "display(Markdown(\"__Fig. 1__ Workflow of the scRNAseq2Targets pipeline. The pipeline starts with two single cell profiles from a control and experimental condition or a bulk RNA-seq dataset with control and perturbation conditions along with a tissue-specific single cell reference. Then it utilizes L1000 data to identify compounds most likely to push diseased cell types to a healthy phenotype as well as identifying cell-type specific membrane targets for targeted cell removal.\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "#load the data\n",
    "{% if data_type.raw_value == 'scRNA-seq' %}\n",
    "\n",
    "{% if data_type.value[0].raw_value == 'Plain Text' %}\n",
    "sc_data_ctrl_file = {{ data_type.value[0].value[0] }}\n",
    "sc_metadata_ctrl_file = {{ data_type.value[0].value[1] }}\n",
    "sc_data_ctrl = read_sc_data(sc_data_ctrl_file, sc_metadata_ctrl_file, 'plain')\n",
    "cell_type_col = {{ data_type.value[1].value[2] }}\n",
    "{% elif data_type.value[1].raw_value == '.mtx' %}\n",
    "## TODO IMPLEMENT .MTX FILE LOADING\n",
    "{% endif %}\n",
    "\n",
    "{% if data_type.value[1].raw_value == 'Plain Text' %}\n",
    "sc_data_pert_file = {{ data_type.value[1].value[0] }}\n",
    "sc_metadata_pert_file = {{ data_type.value[1].value[1] }}\n",
    "sc_data_pert = read_sc_data(sc_data_pert_file, sc_metadata_pert_file, 'plain')\n",
    "\n",
    "cell_type_col = {{ data_type.value[1].value[2] }}\n",
    "{% elif data_type.value[1].raw_value == '.mtx' %}\n",
    "## TODO IMPLEMENT .MTX FILE LOADING\n",
    "{% endif %}\n",
    "\n",
    "{% elif data_type.raw_value == 'bulk RNA-seq' %}\n",
    "\n",
    "bulk_expr_df = read_bulk_data('{{ data_type.value[0].value }}')\n",
    "metadata_df = read_bulk_data('{{ data_type.value[1].value }}')\n",
    "condition_col = '{{ data_type.value[2].value }}'\n",
    "control_cond = '{{ data_type.value[3].value }}'\n",
    "sc_ref = '{{ data_type.value[4].value }}'\n",
    "\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "n_neighbors = {{ n_neighbors.raw_value }}\n",
    "min_dist = {{ min_dist.raw_value }}\n",
    "resolution = {{ resolution.raw_value }}\n",
    "species = '{{ species.raw_value }}'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "try:\n",
    "    lookup = ncbi_genes_lookup(organism='Mammalia/Homo_sapiens' if species == 'human' else 'Mammalia/Mus_musculus')\n",
    "except:\n",
    "    import urllib.request, json \n",
    "    with urllib.request.urlopen(\"https://s3.amazonaws.com/multiomics2paper/public/ncbi_genes_disambiguated.json\") as url:\n",
    "        ncbi_genes_disambiguated = json.load(url)\n",
    "        lookup = ncbi_genes_disambiguated.get\n",
    "{% if data_type.raw_value == 'bulk RNA-seq' %}\n",
    "mamamla_info = pd.read_csv('Mammalia/Homo_sapiens.gene_info.tsv', sep='\\t', index_col=0)\n",
    "syns = list(s.split('|') for s in mamamla_info['dbXrefs'])\n",
    "convert_dict = {}\n",
    "for syn, symbol in zip(syns, list(mamamla_info['Symbol'])):\n",
    "    if len(syn) >= 3 and syn[2].startswith('Ensembl:'):\n",
    "        convert_dict[symbol] = syn[2].split(':')[1]\n",
    "\n",
    "bulk_expr_df.index = bulk_expr_df.index.map(lambda x: lookup(x) if lookup(x) else x)\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "{% if data_type.raw_value == 'bulk RNA-seq' %}\n",
    "# Visualize Bulk Samples\n",
    "Utilize UMAP to visualize bulk quantile and log-normalized bulk data expression data colored by experimental condition.\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% if data_type.raw_value == 'bulk RNA-seq' %}\n",
    "\n",
    "#Visualize Bulk data by UMAP, use adata for convenience\n",
    "adata = sc.AnnData(log2_normalize(qnorm.quantile_normalize(bulk_expr_df)).T.values)\n",
    "adata.var['gene_names'] = bulk_expr_df.index.values\n",
    "adata.obs['samples'] = bulk_expr_df.columns.values\n",
    "adata.obs['condition'] = adata.obs['samples'].map(lambda x: metadata_df.loc[x, condition_col])\n",
    "\n",
    "sc.pp.pca(adata, n_comps=2)\n",
    "sc.pp.neighbors(adata, n_neighbors=n_neighbors, n_pcs=2)\n",
    "sc.tl.umap(adata, min_dist=min_dist)\n",
    "sc.tl.leiden(adata, resolution=resolution)\n",
    "\n",
    "sc.pl.umap(adata, color='condition', legend_fontweight='normal', legend_fontsize=12, size=100, save=f'_bulk_data.png')\n",
    "sc.pl.umap(adata, color='condition', legend_fontweight='normal', legend_fontsize=12, size=100, save=f'_bulk_data.svg', show=False)\n",
    "\n",
    "display(Markdown(f'__Fig. {fig_counter}__. UMAP visualization of quantile and log-normalized bulk data expression data colored by experimental condition.'))\n",
    "display(FileLink('figures/umap_bulk_data.png', result_html_prefix='Download PNG: '))\n",
    "display(FileLink('figures/umap_bulk_data.svg', result_html_prefix='Download SVG: '))\n",
    "fig_counter += 1\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "{% if data_type.raw_value == 'bulk RNA-seq' %}\n",
    "# Deconvolution with InstaPrism\n",
    "InstaPrism [TODO: CITATION] is a deconvolution tool based on the BayesPrism algorithm [TODO: CITATION]. It utilizes a single cell RNA-seq cell type reference for deconvolution into cell type fractions for each bulk sample and also produces cell type specific estimated expression for each bulk sample.\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% if data_type.raw_value == 'bulk RNA-seq' %}\n",
    "#Perform deconvolution with InstaPrism\n",
    "#ref_url: str, bulk_expr: pd.DataFrame, output_dir: str, convert_dict: dict\n",
    "ref_url = f'https://minio.dev.maayanlab.cloud/sc2targets/references/{sc_ref}.rds'\n",
    "bulk_expr_norm = log2_normalize(qnorm.quantile_normalize(bulk_expr_df))\n",
    "with suppress_output():\n",
    "    estimated_frac_df, cell_type_dfs = deconvolution_insta_prism(ref_url, bulk_expr_norm, 'results', convert_dict)\n",
    "shutil.make_archive('deconvolution_res', 'zip', 'results')\n",
    "display(FileLink('deconvolution_res.zip', result_html_prefix='Download deconvolution results: '))\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% if data_type.raw_value == 'bulk RNA-seq' %}\n",
    "# visualize cell type fractions per condition\n",
    "frac_df_viz = estimated_frac_df.copy()\n",
    "frac_df_viz['condition'] = frac_df_viz.index.map(lambda x: metadata_df.loc[x, condition_col])\n",
    "frac_df_viz.reset_index(inplace=True)\n",
    "frac_df_viz.rename(columns={'index': 'sample'}, inplace=True)\n",
    "df_melted = frac_df_viz.melt(id_vars=['sample', 'condition'], var_name='cell_type', value_name='fraction')\n",
    "\n",
    "plt.figure(figsize=(12, 8))\n",
    "sns.boxplot(data=df_melted, x='condition', y='fraction', hue='cell_type', palette='Set2')\n",
    "plt.xlabel('Condition')\n",
    "plt.ylabel('Fraction')\n",
    "plt.legend(title='Cell Type', bbox_to_anchor=(1.05, 1), loc='upper left')\n",
    "plt.xticks(rotation=45)\n",
    "plt.tight_layout()\n",
    "plt.savefig('figures/cell_type_fractions.svg', dpi=300)\n",
    "plt.savefig('figures/cell_type_fractions.png', dpi=300)\n",
    "plt.show()\n",
    "\n",
    "display(Markdown(f'__Fig. {fig_counter}__. Cell type fractions per condition computed using the selected cell type reference and InstaPrism.'))\n",
    "display(FileLink('figures/cell_type_fractions.svg', result_html_prefix='Download SVG: '))\n",
    "display(FileLink('figures/cell_type_fractions.png', result_html_prefix='Download PNG: '))\n",
    "fig_counter += 1\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% if data_type.raw_value == 'bulk RNA-seq' %}\n",
    "\n",
    "pivot_df = df_melted.groupby(['condition', 'cell_type']).mean('fraction').reset_index().pivot(index='cell_type', columns='condition', values='fraction')\n",
    "\n",
    "pert_conditions = list(set([c for c in metadata_df[condition_col].values if c != control_cond]))\n",
    "# Calculate the difference\n",
    "for pert in pert_conditions:\n",
    "    if 'change' in pivot_df.columns:\n",
    "        pivot_df['change'] = pivot_df[pert] - pivot_df[control_cond] + pivot_df['change']\n",
    "    else:\n",
    "        pivot_df['change'] = pivot_df[pert] - pivot_df[control_cond]\n",
    "\n",
    "# Get the absolute change and sort\n",
    "pivot_df['abs_change'] = pivot_df['change'].abs()\n",
    "most_changed_df = pivot_df.sort_values(by='abs_change', ascending=False)\n",
    "\n",
    "# Display the results\n",
    "top_5 = most_changed_df.index.values[:5]\n",
    "display(most_changed_df[['change', 'abs_change']])\n",
    "display(Markdown(f'__Table. {table_counter}__. Cell types ranked by change across control and perturbed conditions.'))\n",
    "table_counter += 1\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% if data_type.raw_value == 'bulk RNA-seq' %}\n",
    "\n",
    "to_merge = []\n",
    "for ct in cell_type_dfs:\n",
    "    ct_name = ct.split('_Z')[0]\n",
    "    if ct_name not in top_5:\n",
    "        continue\n",
    "    ct_df = cell_type_dfs[ct]\n",
    "    ct_df.index =  ct_df.index.map(lambda x: f\"{x}_{ct_name}_{metadata_df.loc[x, condition_col]}\" if x in metadata_df.index else x)\n",
    "    ct_df = ct_df.T \n",
    "    ct_df.index = ct_df.index.map(lambda x: lookup(x) if lookup(x) else x)\n",
    "    to_merge.append(ct_df)\n",
    "bulk_ct_df = pd.concat(to_merge, axis=1)\n",
    "bulk_ct_df\n",
    "\n",
    "adata = sc.AnnData(log2_normalize(qnorm.quantile_normalize(bulk_ct_df)).T.values)\n",
    "adata.var['gene_names'] = bulk_ct_df.index.values\n",
    "adata.obs['samples'] = bulk_ct_df.columns.map(lambda x: x.split('_')[0])\n",
    "adata.obs['cell type'] = bulk_ct_df.columns.map(lambda x: x.split('_')[1])\n",
    "adata.obs['condition'] = bulk_ct_df.columns.map(lambda x: x.split('_')[2])\n",
    "\n",
    "sc.pp.pca(adata, n_comps=2)\n",
    "sc.pp.neighbors(adata, n_neighbors=n_neighbors, n_pcs=2)\n",
    "sc.tl.umap(adata, min_dist=min_dist)\n",
    "sc.tl.leiden(adata, resolution=resolution)\n",
    "\n",
    "sc.pl.umap(adata, color='condition', legend_fontweight='normal', legend_fontsize=12, size=100, save='_condition_cell_type.png')\n",
    "sc.pl.umap(adata, color='condition', legend_fontweight='normal', legend_fontsize=12, size=100, save='_condition_cell_type.svg', show=False)\n",
    "display(Markdown(f'__Fig. {fig_counter}__. UMAP visualization of quantile and log-normalized bulk data expression data colored by experimental condition.'))\n",
    "display(FileLink('figures/umap_condition_cell_type.png', result_html_prefix='Download PNG: '))\n",
    "display(FileLink('figures/umap_condition_cell_type.svg', result_html_prefix='Download SVG: '))\n",
    "fig_counter += 1\n",
    "\n",
    "sc.pl.umap(adata, color='cell type', legend_fontweight='normal', legend_fontsize=12, size=100, save='_cell_type.png')\n",
    "sc.pl.umap(adata, color='cell type', legend_fontweight='normal', legend_fontsize=12, size=100, save='_cell_type.svg', show=False)\n",
    "display(Markdown(f'__Fig. {fig_counter}__. UMAP visualization of quantile and log-normalized bulk data expression data colored by cell type.'))\n",
    "display(FileLink('figures/umap_cell_type.png', result_html_prefix='Download PNG: '))\n",
    "display(FileLink('figures/umap_cell_type.svg', result_html_prefix='Download SVG: '))\n",
    "fig_counter += 1\n",
    "\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "{% if data_type.raw_value == 'scRNA-seq' %}\n",
    "# Cell Type Identification\n",
    "To identify cell types, first cells in each profile are clustered using UMAP [TODO: CITE] and the Leiden algorithm [TODO: CITE]. Canonical cell type makers are sourced from PanglaoDB [TODO: CITE] and are used for enrichment analysis and visualization of inferred cell types using decoupleR [TODO:CITE 36699385].\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% if data_type.raw_value == 'scRNA-seq' %}\n",
    "## Identify cell types if no cell type column is provided\n",
    "sc_data_ctrl = normalize_sc_data(sc_data_ctrl, n_neighbors, min_dist, resolution)\n",
    "sc.pl.umap(sc_data_ctrl, color='leiden', title='scRNA-seq control profile UMAP',\n",
    "           frameon=False, legend_fontweight='normal', legend_fontsize=12, save='_leiden_control.png')\n",
    "sc.pl.umap(sc_data_ctrl, color='leiden', title='scRNA-seq control profile UMAP',\n",
    "           frameon=False, legend_fontweight='normal', legend_fontsize=12, save='_leiden_control.svg', show=False)\n",
    "display(Markdown(f'__Fig. {fig_counter}__. UMAP visualization of scRNA-seq control data colored by leiden cluster.'))\n",
    "display(FileLink('figures/umap_leiden_control.png', result_html_prefix='Download PNG: '))\n",
    "display(FileLink('figures/umap_leiden_control.svg', result_html_prefix='Download SVG: '))\n",
    "fig_counter += 1\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% if data_type.raw_value == 'scRNA-seq' %}\n",
    "## Identify cell types if no cell type column is provided\n",
    "sc_data_pert = normalize_sc_data(sc_data_pert, n_neighbors, min_dist, resolution)\n",
    "sc.pl.umap(sc_data_pert, color='leiden', title='scRNA-seq perturbation profile UMAP',\n",
    "           frameon=False, legend_fontweight='normal', legend_fontsize=12, save='_leiden_pert.png')\n",
    "sc.pl.umap(sc_data_pert, color='leiden', title='scRNA-seq perturbation profile UMAP',\n",
    "           frameon=False, legend_fontweight='normal', legend_fontsize=12, save='_leiden_pert.svg', show=False)\n",
    "display(Markdown(f'__Fig. {fig_counter}__. UMAP visualization of scRNA-seq perturbation data colored by leiden cluster.'))\n",
    "display(FileLink('figures/umap_leiden_pert.png', result_html_prefix='Download PNG: '))\n",
    "display(FileLink('figures/umap_leiden_pert.svg', result_html_prefix='Download SVG: '))\n",
    "fig_counter += 1\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% if data_type.raw_value == 'scRNA-seq' %}\n",
    "markers = pd.read_csv('https://minio.dev.maayanlab.cloud/sc2targets/PanglaoDB_markers_27_Mar_2020.tsv', sep='\\t')\n",
    "# Filter by canonical_marker and human\n",
    "markers['human'] = markers['species'].str.contains('Hs')\n",
    "markers['mouse'] = markers['species'].str.contains('Mm')\n",
    "markers = markers[markers[species] & (markers['canonical marker'] == 1) & (markers[f'sensitivity_{species}'] > 0.5)]\n",
    "markers = markers[~markers.duplicated(['cell type', 'official gene symbol'])]\n",
    "markers.rename(columns={'cell type': 'cell_type', 'official gene symbol': 'genesymbol'}, inplace=True)\n",
    "if species == 'mouse':\n",
    "    markers['genesymbol'] = markers['genesymbol'].map(lambda x: x.lower().capitalize())\n",
    "markers\n",
    "\n",
    "\n",
    "sc_data_pert.var.set_index('gene_names', inplace=True)\n",
    "sc_data_ctrl.var.set_index('gene_names', inplace=True)\n",
    "\n",
    "dc.run_ora(\n",
    "    mat=sc_data_pert,\n",
    "    net=markers,\n",
    "    source='cell_type',\n",
    "    target='genesymbol',\n",
    "    min_n=3,\n",
    "    verbose=True,\n",
    "    use_raw=False\n",
    ")\n",
    "\n",
    "dc.run_ora(\n",
    "    mat=sc_data_ctrl,\n",
    "    net=markers,\n",
    "    source='cell_type',\n",
    "    target='genesymbol',\n",
    "    min_n=3,\n",
    "    verbose=True,\n",
    "    use_raw=False\n",
    ")\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% if data_type.raw_value == 'scRNA-seq' %}\n",
    "acts_ctrl = dc.get_acts(sc_data_ctrl, obsm_key='ora_estimate')\n",
    "acts_pert = dc.get_acts(sc_data_pert, obsm_key='ora_estimate')\n",
    "sc_ctrl_cell_types = dc.rank_sources_groups(acts_ctrl, groupby='leiden', reference='rest', method='t-test_overestim_var')\n",
    "sc_pert_cell_types = dc.rank_sources_groups(acts_pert, groupby='leiden', reference='rest', method='t-test_overestim_var')\n",
    "n_ctypes = 3\n",
    "ctypes_dict_ctrl = sc_ctrl_cell_types.groupby('group').head(n_ctypes).groupby('group')['names'].apply(lambda x: list(x)).to_dict()\n",
    "ctypes_dict_pert = sc_pert_cell_types.groupby('group').head(n_ctypes).groupby('group')['names'].apply(lambda x: list(x)).to_dict()\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% if data_type.raw_value == 'scRNA-seq' %}\n",
    "\n",
    "sc.pl.matrixplot(acts_ctrl, ctypes_dict_ctrl, 'leiden', dendrogram=True, standard_scale='var',\n",
    "                 colorbar_title='Z-scaled scores', cmap='RdBu_r', save='cell_types_ctrl.png')\n",
    "sc.pl.matrixplot(acts_ctrl, ctypes_dict_ctrl, 'leiden', dendrogram=True, standard_scale='var',\n",
    "                 colorbar_title='Z-scaled scores', cmap='RdBu_r', save='cell_types_ctrl.svg', show=False)\n",
    "display(Markdown(f'__Fig. {fig_counter}__. Matrix visualization of scRNA-seq leiden cluster cell type enrichments for the control profile.'))\n",
    "display(FileLink('figures/matrixplot_cell_types_ctrl.png', result_html_prefix='Download PNG: '))\n",
    "display(FileLink('figures/matrixplot_cell_types_ctrl.svg', result_html_prefix='Download SVG: '))\n",
    "fig_counter += 1\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% if data_type.raw_value == 'scRNA-seq' %}\n",
    "\n",
    "sc.pl.matrixplot(acts_pert, ctypes_dict_pert, 'leiden', dendrogram=True, standard_scale='var',\n",
    "                 colorbar_title='Z-scaled scores', cmap='RdBu_r', save='cell_types_pert.png')\n",
    "sc.pl.matrixplot(acts_pert, ctypes_dict_pert, 'leiden', dendrogram=True, standard_scale='var',\n",
    "                 colorbar_title='Z-scaled scores', cmap='RdBu_r', save='cell_types_pert.svg', show=False)\n",
    "display(Markdown(f'__Fig. {fig_counter}__. Matrix visualization of scRNA-seq leiden cluster cell type enrichments for the perturbation profile.'))\n",
    "display(FileLink('figures/matrixplot_cell_types_pert.png', result_html_prefix='Download PNG: '))\n",
    "display(FileLink('figures/matrixplot_cell_types_pert.svg', result_html_prefix='Download SVG: '))\n",
    "\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% if data_type.raw_value == 'scRNA-seq' %}\n",
    "\n",
    "annotation_dict_ctrl = sc_ctrl_cell_types.groupby('group').head(1).set_index('group')['names'].to_dict()\n",
    "annotation_dict_pert = sc_pert_cell_types.groupby('group').head(1).set_index('group')['names'].to_dict()\n",
    "sc_data_ctrl.obs['cell_type'] = [annotation_dict_ctrl[clust] for clust in sc_data_ctrl.obs['leiden']]\n",
    "sc_data_pert.obs['cell_type'] = [annotation_dict_pert[clust] for clust in sc_data_pert.obs['leiden']]\n",
    "overlapping_cell_types = list(set(annotation_dict_ctrl.values()).intersection(set(annotation_dict_pert.values())))\n",
    "display(Markdown(f\"Overlapping cell types for downstream analysis: {', '.join(overlapping_cell_types)}\"))\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% if data_type.raw_value == 'scRNA-seq' %}\n",
    "sc.pl.umap(sc_data_ctrl, color='cell_type', save='_cell_types_ctrl.png')\n",
    "sc.pl.umap(sc_data_ctrl, color='cell_type', save='_cell_types_ctrl.svg', show=False)\n",
    "display(Markdown(f'__Fig. {fig_counter}__. UMAP visualization of scRNA-seq cell types for the control profile.'))\n",
    "display(FileLink('figures/umap_cell_types_ctrl.png', result_html_prefix='Download PNG: '))\n",
    "display(FileLink('figures/umap_cell_types_ctrl.svg', result_html_prefix='Download SVG: '))\n",
    "fig_counter += 1\n",
    "\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% if data_type.raw_value == 'scRNA-seq' %}\n",
    "sc.pl.umap(sc_data_pert, color='cell_type', save='_cell_types_pert.png')\n",
    "sc.pl.umap(sc_data_pert, color='cell_type', save='_cell_types_pert.svg', show=False)\n",
    "display(Markdown(f'__Fig. {fig_counter}__. UMAP visualization of scRNA-seq cell types for the perturbation profile.'))\n",
    "display(FileLink('figures/umap_cell_types_pert.png', result_html_prefix='Download PNG: '))\n",
    "display(FileLink('figures/umap_cell_types_pert.svg', result_html_prefix='Download SVG: '))\n",
    "fig_counter += 1\n",
    "\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% if data_type.raw_value == 'scRNA-seq' %}\n",
    "\n",
    "## get three representative samples per overlapping cell type\n",
    "data_pert = []\n",
    "sample_names_pert = []\n",
    "\n",
    "data_ctrl = []\n",
    "sample_names_ctrl = []\n",
    "# get indices of shared genes for each mat\n",
    "\n",
    "for ct in overlapping_cell_types:\n",
    "    ## extract counts from samples matching the cell type:\n",
    "    ct_cells = sc_data_pert.X.transpose()[:, sc_data_pert.obs['cell_type'] == ct]\n",
    "    n_cells = ct_cells.shape[1]\n",
    "    for i in range(3):\n",
    "        rand_idx = np.random.choice(list(range(n_cells)), n_cells // 3, replace=False)\n",
    "        ct_cells_rand_samp = ct_cells[:, rand_idx].sum(axis=1)\n",
    "        data_pert.append(ct_cells_rand_samp)\n",
    "        sample_names_pert.append(f'pert_{ct}_{i + 1}')\n",
    "    \n",
    "    ct_cells = sc_data_ctrl.X.transpose()[:, sc_data_ctrl.obs['cell_type'] == ct]\n",
    "    n_cells = ct_cells.shape[1]\n",
    "    for i in range(3):\n",
    "        rand_idx = np.random.choice(list(range(n_cells)), n_cells // 3, replace=False)\n",
    "        ct_cells_rand_samp = ct_cells[:, rand_idx].sum(axis=1)\n",
    "        data_ctrl.append(ct_cells_rand_samp)\n",
    "        sample_names_ctrl.append(f'ctrl_{ct}_{i + 1}')\n",
    "\n",
    "sim_bulk_pert = pd.DataFrame(data=data_pert, index=sample_names_pert, columns=sc_data_pert.var_names).T\n",
    "sim_bulk_ctrl = pd.DataFrame(data=data_ctrl, index=sample_names_ctrl, columns=sc_data_ctrl.var_names).T\n",
    "bulk_ct_df = pd.merge(sim_bulk_pert, sim_bulk_ctrl, left_index=True, right_index=True)\n",
    "bulk_ct_df.to_csv('results/bulk_ct_df.csv')\n",
    "display(bulk_ct_df)\n",
    "display(Markdown(f'__Table. {table_counter}__. Three representative samples per overlapping cell type in the control and perturbation profiles.'))\n",
    "display(FileLink('results/bulk_ct_df.csv', result_html_prefix='Download CSV: '))\n",
    "table_counter += 1\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "targets = {}\n",
    "if species == 'mouse':\n",
    "    lookup = ncbi_genes_lookup(organism='Mammalia/Mus_musculus')\n",
    "    bgs = {'ARCHS4': 'Mammalia/Mus_musculus/archs4-gene-stats.tsv', 'Tabula Muris': 'Mammalia/Mus_musculus/tabula-muris-gene-stats.tsv' }\n",
    "else:\n",
    "    lookup = ncbi_genes_lookup()\n",
    "    bgs = {'GTEx': 'gtex-gene-stats.tsv', 'ARCHS4': 'archs4-gene-stats.tsv', 'TS': 'ts_10x_cell-ontology-class_donors_tissue-labels_v1.tsv'}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "# Target Identification\n",
    "To identify highly expressed membrane targets on perturbed cell types, we utilize TargetRanger [TODO: CITE], a collection of heathy tissue and cell type atlases. TargetRanger works by finding genes highly expressed in perturbed cell types that are lowly expressed across healthy cell types and tissues. Additionally, we filter targets identified for the control profiles of each cell type to uniquely target diseased cell types."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_targets(rna_df, bg, targets):\n",
    "    df_bg_stats = pd.read_csv(f\"https://appyters.maayanlab.cloud/storage/Tumor_Gene_Target_Screener/{bgs[bg]}\", sep='\\t', index_col=[0,1])\n",
    "    df_bg_genes = df_bg_stats.unstack().index.map(lambda idx: lookup(idx.partition('.')[0]))\n",
    "    df_bg_stats = df_bg_stats.unstack().groupby(df_bg_genes, observed=True).median().stack()\n",
    "    df_bg_expr = df_bg_stats.loc[(slice(None), ['25%', '50%', '75%']), :].unstack()\n",
    "    common_index = list(set(rna_df.index) & set(df_bg_expr.index))\n",
    "    expr_df = rna_df.loc[common_index, :]\n",
    "    index_name = expr_df.index.name\n",
    "    expr_df.reset_index(inplace=True)\n",
    "    expr_df.drop_duplicates(subset=index_name, inplace=True)\n",
    "    expr_df.set_index(index_name, inplace=True, drop=True)\n",
    "    target_distribution = df_bg_expr.loc[common_index, :].median(axis=1)\n",
    "    df_expr_norm = qnorm.quantile_normalize(expr_df.loc[common_index, :], target=target_distribution)\n",
    "    df_bg_expr_norm = qnorm.quantile_normalize(df_bg_expr.loc[common_index, :], target=target_distribution)\n",
    "    for ct in tqdm(overlapping_cell_types):\n",
    "        if ct not in targets: targets[ct] = {}\n",
    "        ct_pert_samples = list(filter(lambda x: ct in x and 'pert' in x, bulk_ct_df.columns))\n",
    "        ct_ctrl_samples = list(filter(lambda x: ct in x and 'ctrl' in x, bulk_ct_df.columns))\n",
    "        with suppress_output():\n",
    "            df_bg_expr_norm.columns = df_bg_expr_norm.columns.to_flat_index().map(lambda s: ', '.join(s))\n",
    "            dge_pert = ttest_differential_expression(\n",
    "                df_bg_expr_norm, df_expr_norm[ct_pert_samples],\n",
    "                log2norm=False,\n",
    "            )\n",
    "            dge_ctrl = ttest_differential_expression(\n",
    "                df_bg_expr_norm, df_expr_norm[ct_ctrl_samples],\n",
    "                log2norm=False,\n",
    "            )\n",
    "            overlapping_pert_norm = set(dge_pert[(dge_pert['AdjPval'] < 0.01) & (dge_pert['Statistic'] > 0)].sort_values('Statistic', ascending=False).index.values).difference(set(dge_ctrl[(dge_ctrl['AdjPval'] < 0.01) & (dge_ctrl['Statistic'] > 0)].sort_values('Statistic', ascending=False).index.values))\n",
    "            targets[ct][bg] = list(dge_pert[(dge_pert['AdjPval'] < 0.01) & (dge_pert['Statistic'] > 0) & (~dge_pert.index.isin(overlapping_pert_norm))].sort_values('Statistic', ascending=False).index.values)\n",
    "\n",
    "def find_targets_multiple_pert(rna_df, bg, targets, control_cond, pert_conditions):\n",
    "    df_bg_stats = pd.read_csv(f\"/Users/giacomomarino/GeneRangerFigures/datasets/{bgs[bg]}\", sep='\\t', index_col=[0,1])\n",
    "    df_bg_genes = df_bg_stats.unstack().index.map(lambda idx: lookup(idx.partition('.')[0]))\n",
    "    df_bg_stats = df_bg_stats.unstack().groupby(df_bg_genes, observed=True).median().stack()\n",
    "    df_bg_expr = df_bg_stats.loc[(slice(None), ['25%', '50%', '75%']), :].unstack()\n",
    "    common_index = list(set(rna_df.index) & set(df_bg_expr.index))\n",
    "    expr_df = rna_df.loc[common_index, :]\n",
    "    index_name = expr_df.index.name\n",
    "    if index_name == None:\n",
    "        expr_df.index.name = 'gene_name'\n",
    "        index_name = 'gene_name'\n",
    "    expr_df.reset_index(inplace=True)\n",
    "    expr_df.drop_duplicates(subset=index_name, inplace=True)\n",
    "    expr_df.set_index(index_name, inplace=True, drop=True)\n",
    "    target_distribution = df_bg_expr.loc[common_index, :].median(axis=1)\n",
    "    df_expr_norm = qnorm.quantile_normalize(expr_df.loc[common_index, :], target=target_distribution)\n",
    "    df_bg_expr_norm = qnorm.quantile_normalize(df_bg_expr.loc[common_index, :], target=target_distribution)\n",
    "    df_bg_expr_norm.columns = df_bg_expr_norm.columns.to_flat_index().map(lambda s: ', '.join(s))\n",
    "    for ct in tqdm(top_5):\n",
    "        ct_ctrl_samples = list(filter(lambda x: ct in x and control_cond in x, bulk_ct_df.columns))\n",
    "        with suppress_output():\n",
    "            dge_ctrl = ttest_differential_expression(\n",
    "                df_bg_expr_norm, df_expr_norm[ct_ctrl_samples],\n",
    "                log2norm=False,\n",
    "            )\n",
    "            for pert in pert_conditions:\n",
    "                if f\"{ct}-{pert}\" not in targets: targets[f\"{ct}-{pert}\"] = {}\n",
    "                ct_pert_samples = list(filter(lambda x: ct in x and pert in x, bulk_ct_df.columns))\n",
    "                dge_pert = ttest_differential_expression(\n",
    "                    df_bg_expr_norm, df_expr_norm[ct_pert_samples],\n",
    "                    log2norm=False,\n",
    "                )\n",
    "                overlapping_pert_norm = set(dge_pert[(dge_pert['AdjPval'] < 0.01) & (dge_pert['Statistic'] > 0)].sort_values('Statistic', ascending=False).index.values).difference(set(dge_ctrl[(dge_ctrl['AdjPval'] < 0.01) & (dge_ctrl['Statistic'] > 0)].sort_values('Statistic', ascending=False).index.values))\n",
    "                targets[f\"{ct}-{pert}\"][bg] = list(dge_pert[(dge_pert['AdjPval'] < 0.01) & (dge_pert['Statistic'] > 0) & (~dge_pert.index.isin(overlapping_pert_norm))].sort_values('Statistic', ascending=False).index.values)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "if species == 'human': \n",
    "{% if data_type.raw_value == 'scRNA-seq' %}\n",
    "    print('Finding Targets using ARCHS4')\n",
    "    find_targets(bulk_ct_df, 'ARCHS4', targets)  \n",
    "    print('Finding Targets using GTEx')\n",
    "    find_targets(bulk_ct_df, 'GTEx', targets)\n",
    "    print('Finding Targets using Tabula Sapiens')\n",
    "    find_targets(bulk_ct_df, 'TS', targets)\n",
    "{% elif data_type.raw_value == 'bulk RNA-seq' %}\n",
    "    print('Finding Targets using ARCHS4')\n",
    "    find_targets_multiple_pert(bulk_ct_df, 'ARCHS4', targets, control_cond, pert_conditions)\n",
    "    print('Finding Targets using GTEx')\n",
    "    find_targets_multiple_pert(bulk_ct_df, 'GTEx', targets, control_cond, pert_conditions)\n",
    "    print('Finding Targets using Tabula Sapiens')\n",
    "    find_targets_multiple_pert(bulk_ct_df, 'TS', targets, control_cond, pert_conditions)\n",
    "{% endif %}\n",
    "elif species == 'mouse':\n",
    "{% if data_type.raw_value == 'scRNA-seq' %}\n",
    "    print('Finding Targets using ARCHS4')\n",
    "    find_targets(bulk_ct_df, 'ARCHS4', targets)\n",
    "    print('Finding Targets using Tabula Muris')\n",
    "    find_targets(bulk_ct_df, 'Tabula Muris', targets)\n",
    "{% elif data_type.raw_value == 'bulk RNA-seq' %}\n",
    "    print('Finding Targets using ARCHS4')\n",
    "    find_targets_multiple_pert(bulk_ct_df, 'ARCHS4', targets, control_cond, pert_conditions)\n",
    "    print('Finding Targets using Tabula Muris')\n",
    "    find_targets_multiple_pert(bulk_ct_df, 'Tabula Muris', targets, control_cond, pert_conditions)\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "proteins = pd.read_csv('https://appyters.maayanlab.cloud/storage/Tumor_Gene_Target_Screener/surfaceome.csv')\n",
    "membrane_proteins = list(proteins['genename'].dropna().values)\n",
    "\n",
    "if species == 'human':\n",
    "    sc.set_figure_params(scanpy=False)\n",
    "    YlGnBu = cm.get_cmap('YlGnBu_r', 8)\n",
    "    cmap = {\"None\":YlGnBu(0), \"ARCHS4\": YlGnBu(1), \"GTEx\":YlGnBu(2), \"TS\": YlGnBu(4), \"ARCHS4-GTEx\":YlGnBu(3),  \"ARCHS4-TS\": YlGnBu(5), \"GTEx-TS\": YlGnBu(6), \"All\": YlGnBu(7)}\n",
    "\n",
    "    top_targets_n = 100\n",
    "    target_list = []\n",
    "    for ct in targets:\n",
    "        for bg in targets[ct]:\n",
    "            target_list.extend(list(filter(lambda g: g in membrane_proteins, list(targets[ct][bg])[:top_targets_n])))\n",
    "    if len(set(target_list)) < 100:\n",
    "        top_targets_n = 500\n",
    "        for cluster in targets:\n",
    "            for bg in targets[cluster]:\n",
    "                target_list.extend(list(filter(lambda g: g in membrane_proteins, list(targets[cluster][bg])[:top_targets_n])))\n",
    "    \n",
    "    data1, data2, data3 = [], [], []\n",
    "    similarity = []\n",
    "    target_list = list(set(target_list))\n",
    "\n",
    "    for gene in target_list:\n",
    "        a = [1 if gene in list(targets[c]['ARCHS4'])[:top_targets_n] else 0 for c in list(targets.keys())]\n",
    "        g = [2 if gene in list(targets[c]['GTEx'])[:top_targets_n] else 0 for c in list(targets.keys())]\n",
    "        l = [4 if gene in list(targets[c]['TS'])[:top_targets_n] else 0 for c in list(targets.keys())]\n",
    "        data1.append(a)\n",
    "        data2.append(g)\n",
    "        data3.append(l)\n",
    "        similarity.append(np.dot(np.dot(np.array(a), np.array(g)), np.array(l)))\n",
    "\n",
    "    data = np.add(np.add(data1, data2), data3)\n",
    "\n",
    "    membrane_target_mat = pd.DataFrame(data)\n",
    "    membrane_target_mat.columns = list(targets.keys())\n",
    "    membrane_target_mat.index = target_list\n",
    "\n",
    "    membrane_target_mat['count'] = membrane_target_mat.sum(axis=1)\n",
    "    membrane_target_mat = membrane_target_mat[membrane_target_mat['count'] >= 7]\n",
    "    membrane_target_mat = membrane_target_mat.rename_axis('Membrane Target').sort_values(by = ['count', 'Membrane Target'], ascending = [False, True]).drop('count', axis=1)\n",
    "\n",
    "    h = membrane_target_mat.shape[0]\n",
    "    \n",
    "    g = sns.clustermap(membrane_target_mat, figsize=(4,0.3*h+2*(h<15)), cmap=YlGnBu, cbar_pos=None, dendrogram_ratio=0.1-(h<40)*0.01*(h-30), row_cluster=False, xticklabels=True, yticklabels=True)\n",
    "    g.ax_row_dendrogram.legend(handles=[Rectangle((0, 0), 0, 0, color=val, label=key) for key, val in cmap.items()],\n",
    "                                    title='Background', loc='upper right')\n",
    "    plt.grid(False)\n",
    "    plt.savefig('figures/membrane_targets.svg', dpi=300)\n",
    "    plt.savefig('figures/membrane_targets.png', dpi=300)\n",
    "    plt.show()\n",
    "    display(Markdown(f'__Fig. {fig_counter}__. Membrane Targets for perturbed cell types, filtered by genes upregulated in control cell types compared to healthy background atlases.'))\n",
    "    display(FileLink('figures/membrane_targets.png', result_html_prefix='Download PNG: '))\n",
    "    display(FileLink('figures/membrane_targets.svg', result_html_prefix='Download SVG: '))\n",
    "    fig_counter += 1\n",
    "\n",
    "    \n",
    "elif species == 'mouse':\n",
    "    YlGnBu = cm.get_cmap('YlGnBu_r', 4)\n",
    "    cmap = {\"None\": YlGnBu(0), \"ARCHS4\": YlGnBu(1), \"Tabula Muris\":YlGnBu(2), \"Both\": YlGnBu(3)}\n",
    "\n",
    "    top_targets_n = 100\n",
    "    target_list = []\n",
    "    for ct in targets:\n",
    "        for bg in targets[ct]:\n",
    "            target_list.extend(list(filter(lambda g: g.upper() in membrane_proteins, list(targets[ct][bg]))))\n",
    "            #target_list.extend(list(targets[ct][bg])[:20])\n",
    "\n",
    "    data1, data2 = [], []\n",
    "    similarity = []\n",
    "    target_list = list(set(target_list))\n",
    "\n",
    "    for gene in target_list:\n",
    "        a = [1 if gene in targets[c]['ARCHS4'] else 0 for c in overlapping_cell_types]\n",
    "        g = [2 if gene in targets[c]['Tabula Muris'] else 0 for c in overlapping_cell_types]\n",
    "        data1.append(a)\n",
    "        data2.append(g)\n",
    "        similarity.append(np.dot(np.array(a), np.array(g)))\n",
    "\n",
    "    data = np.add(data1, data2)\n",
    "\n",
    "    membrane_target_mat = pd.DataFrame(data)\n",
    "    membrane_target_mat.columns = overlapping_cell_types\n",
    "    membrane_target_mat.index = target_list\n",
    "    \n",
    "    membrane_target_mat['count'] = membrane_target_mat.sum(axis=1)\n",
    "    membrane_target_mat = membrane_target_mat[membrane_target_mat['count'] >= 3]\n",
    "    \n",
    "    membrane_target_mat = membrane_target_mat.rename_axis('Membrane Target').sort_values(by = ['count', 'Membrane Target'], ascending = [False, True]).drop('count', axis=1)\n",
    "\n",
    "    h = membrane_target_mat.shape[0]\n",
    "    membrane_target_mat = membrane_target_mat.astype(float)\n",
    "    \n",
    "    g = sns.clustermap(membrane_target_mat, figsize=(4,0.3*h+2*(h<15)), cmap=YlGnBu, cbar_pos=None, dendrogram_ratio=0.1-(h<40)*0.01*(h-30), row_cluster=False, xticklabels=True, yticklabels=True)\n",
    "    g.ax_row_dendrogram.legend(handles=[Rectangle((0, 0), 0, 0, color=val, label=key) for key, val in cmap.items()],\n",
    "                                    title='Background', loc='upper right')\n",
    "    plt.grid(False)\n",
    "    plt.savefig('figures/membrane_targets.svg', dpi=300)\n",
    "    plt.savefig('figures/membrane_targets.png', dpi=300)\n",
    "    plt.show()\n",
    "    display(Markdown(f'__Fig. {fig_counter}__. Membrane Targets for perturbed cell types, filtered by genes upregulated in control cell types compared to healthy background atlases.'))\n",
    "    display(FileLink('figures/membrane_targets.png', result_html_prefix='Download PNG: '))\n",
    "    display(FileLink('figures/membrane_targets.svg', result_html_prefix='Download SVG: '))\n",
    "    fig_counter += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% if data_type.raw_value == 'scRNA-seq' %}\n",
    "data = []\n",
    "for ct in overlapping_cell_types:\n",
    "    ct_pert_samples = list(filter(lambda x: ct in x and 'pert' in x, bulk_ct_df.columns))\n",
    "    ct_ctrl_samples = list(filter(lambda x: ct in x and 'ctrl' in x, bulk_ct_df.columns))\n",
    "    if len(ct_pert_samples) > 0 and len(ct_ctrl_samples) > 0:\n",
    "        diff_expr_ct = ttest_differential_expression(bulk_ct_df[ct_ctrl_samples], bulk_ct_df[ct_pert_samples], log2norm=False)\n",
    "        up_genes = list(diff_expr_ct[(diff_expr_ct['Pval'] < 0.01) & (diff_expr_ct['Statistic'] > 0)].sort_values('Statistic', ascending=False).index.values)\n",
    "        down_genes = list(diff_expr_ct[(diff_expr_ct['Pval'] < 0.01) & (diff_expr_ct['Statistic'] < 0)].sort_values('Statistic', ascending=True).index.values)\n",
    "        desc, link = get_sigcom_link({'up_entities': up_genes, 'down_entities': down_genes, 'description': ct})\n",
    "        data.append([ct, link])\n",
    "\n",
    "compound_df = pd.DataFrame(data, columns=['cell_type','link'])\n",
    "compound_df.set_index('cell_type', inplace=True)\n",
    "compound_df.to_csv('results/sigcom_links.csv')\n",
    "compound_df = compound_df.style.format({'link': lambda url: f'<a href=\"{url}\" rel=\"noopener noreferrer\" target=\"_blank\">SigCom LINCS</a>'})\n",
    "display(compound_df)\n",
    "display(Markdown(f'__Table. {table_counter}__. SigCom LINCS analysis links for overlapping cell types in control and perturbation scRNA-seq profiles.'))\n",
    "display(FileLink('results/sigcom_links.csv', result_html_prefix='Download CSV: '))\n",
    "table_counter += 1\n",
    "{% elif data_type.raw_value == 'bulk RNA-seq' %}\n",
    "data = []\n",
    "for ct in tqdm(top_5):\n",
    "    ct_ctrl_samples = list(filter(lambda x: ct in x and control_cond in x, bulk_ct_df.columns))\n",
    "    for pert in pert_conditions:\n",
    "        ct_pert_samples = list(filter(lambda x: ct in x and pert in x, bulk_ct_df.columns))\n",
    "        diff_expr_ct = ttest_differential_expression(bulk_ct_df[ct_ctrl_samples], bulk_ct_df[ct_pert_samples], log2norm=False)\n",
    "        up_genes = list(diff_expr_ct[(diff_expr_ct['Pval'] < 0.01) & (diff_expr_ct['Statistic'] > 0)].sort_values('Statistic', ascending=False).index.values)\n",
    "        down_genes = list(diff_expr_ct[(diff_expr_ct['Pval'] < 0.01) & (diff_expr_ct['Statistic'] < 0)].sort_values('Statistic', ascending=True).index.values)\n",
    "        if len(up_genes) >= 5 and len(down_genes) >= 5:\n",
    "            desc, link = get_sigcom_link({'up_entities': up_genes, 'down_entities': down_genes, 'description': ct})\n",
    "            data.append([ct, pert, link])\n",
    "\n",
    "compound_df = pd.DataFrame(data, columns=['cell_type', 'perturbation', 'link'])\n",
    "compound_df.set_index('cell_type', inplace=True)\n",
    "compound_df.to_csv('results/sigcom_links.csv')\n",
    "compound_df = compound_df.style.format({'link': lambda url: f'<a href=\"{url}\" rel=\"noopener noreferrer\" target=\"_blank\">SigCom LINCS</a>'})\n",
    "display(compound_df)\n",
    "display(Markdown(f'__Table. {table_counter}__. SigCom LINCS analysis links for top 5 changed perturbed cell types.'))\n",
    "display(FileLink('results/sigcom_links.csv', result_html_prefix='Download CSV: '))\n",
    "table_counter += 1\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# References"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "\n",
    "TODO: ADD ORDERING/ change based on species and uploaded data type\n",
    "\n",
    "\n",
    "Butler A, Hoffman P, Smibert P, Papalexi E, Satija R. Integrating single-cell transcriptomic data across different conditions, technologies, and species. Nat Biotechnol. 2018 Jun;36(5):411-420. doi: 10.1038/nbt.4096. \n",
    "\n",
    "Hu M, Chikina M. InstaPrism: an R package for fast implementation of BayesPrism. Bioinformatics. 2024 Jul 1;40(7):btae440.\n",
    "\n",
    "Chu T, Wang Z, Pe'er D, Danko CG. Cell type and gene expression deconvolution with BayesPrism enables Bayesian integrative analysis across bulk and single-cell RNA sequencing in oncology. Nat Cancer. 2022 Apr;3(4):505-517.\n",
    "\n",
    "Marino GB, Ngai M, Clarke DJB, Fleishman RH, Deng EZ, Xie Z, Ahmed N, Ma'ayan A. GeneRanger and TargetRanger: processed gene and protein expression levels across cells and tissues for target discovery. Nucleic Acids Res. 2023 Jul 5;51(W1):W213-W224.\n",
    "\n",
    "Evangelista JE, Clarke DJB, Xie Z, Lachmann A, Jeon M, Chen K, Jagodnik KM, Jenkins SL, Kuleshov MV, Wojciechowicz ML, Schürer SC, Medvedovic M, Ma'ayan A. SigCom LINCS: data and metadata search engine for a million gene expression signatures. Nucleic Acids Res. 2022 Jul 5;50(W1):W697-W709.\n",
    "\n",
    "Lachmann A, Torre D, Keenan AB, Jagodnik KM, Lee HJ, Wang L, Silverstein MC, Ma'ayan A. Massive mining of publicly available RNA-seq data from human and mouse. Nat Commun. 2018 Apr 10;9(1):1366.\n",
    "\n",
    "GTEx Consortium. The Genotype-Tissue Expression (GTEx) project. Nat Genet. 2013 Jun;45(6):580-5. doi: 10.1038/ng.2653. \n",
    "\n",
    "Tabula Sapiens Consortium. The Tabula Sapiens: A multiple-organ, single-cell transcriptomic atlas of humans. Science. 2022 May 13;376(6594):eabl4896. doi: 10.1126/science.abl4896.\n",
    "\n",
    "Tabula Muris Consortium. Single-cell transcriptomics of 20 mouse organs creates a Tabula Muris. Nature. 2018 Oct;562(7727):367-372. doi: 10.1038/s41586-018-0590-4. \n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
